
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  
  <!-- Licensed under the Apache 2.0 License -->
  <link rel="stylesheet" type="text/css" href="_static/fonts/open-sans/stylesheet.css" />
  <!-- Licensed under the SIL Open Font License -->
  <link rel="stylesheet" type="text/css" href="_static/fonts/source-serif-pro/source-serif-pro.css" />
  <link rel="stylesheet" type="text/css" href="_static/css/bootstrap.min.css" />
  <link rel="stylesheet" type="text/css" href="_static/css/bootstrap-theme.min.css" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
    <title>10. Bash lesson 2 &#8212; 2019-04-11-gwu 0.0.1 documentation</title>
    <link rel="stylesheet" href="_static/guzzle.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="11. Working with Bioconda." href="bioconda_lesson.html" />
    <link rel="prev" title="9. Introducing the Shell" href="bash_lesson_1.html" />
  
   

  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="bioconda_lesson.html" title="11. Working with Bioconda."
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="bash_lesson_1.html" title="9. Introducing the Shell"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="toc.html">2019-04-11-gwu 0.0.1 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="container-wrapper">

      <div id="mobile-toggle">
        <a href="#"><span class="glyphicon glyphicon-align-justify" aria-hidden="true"></span></a>
      </div>
  <div id="left-column">
    <div class="sphinxsidebar">
<div class="sidebar-block">
  <div class="sidebar-wrapper">
    <h2>Contents</h2>
    <div class="sidebar-localtoc">
      <ul>
<li><a class="reference internal" href="#">10. Bash lesson 2</a><ul>
<li><a class="reference internal" href="#working-with-files">10.1. Working with Files</a><ul>
<li><a class="reference internal" href="#our-data-set-fastq-files">10.1.1. Our data set: FASTQ files</a></li>
<li><a class="reference internal" href="#wild-cards">10.1.2. Wild cards</a></li>
</ul>
</li>
<li><a class="reference internal" href="#command-history">10.2. Command History</a></li>
<li><a class="reference internal" href="#examining-files">10.3. Examining Files</a></li>
<li><a class="reference internal" href="#details-on-the-fastq-format">10.4. Details on the FASTQ format</a></li>
<li><a class="reference internal" href="#creating-moving-copying-and-removing">10.5. Creating, moving, copying, and removing</a><ul>
<li><a class="reference internal" href="#copying-files">10.5.1. Copying Files</a></li>
<li><a class="reference internal" href="#creating-directories">10.5.2. Creating Directories</a></li>
<li><a class="reference internal" href="#moving-renaming">10.5.3. Moving / Renaming</a></li>
<li><a class="reference internal" href="#file-permissions">10.5.4. File Permissions</a></li>
<li><a class="reference internal" href="#removing">10.5.5. Removing</a></li>
</ul>
</li>
<li><a class="reference internal" href="#searching-files">10.6. Searching files</a></li>
<li><a class="reference internal" href="#redirecting-output">10.7. Redirecting output</a></li>
<li><a class="reference internal" href="#file-manipulation-and-more-practice-with-pipes">10.8. File manipulation and more practice with pipes</a><ul>
<li><a class="reference internal" href="#how-many-of-the-read-libraries-are-paired-end">10.8.1. How many of the read libraries are paired end?</a></li>
<li><a class="reference internal" href="#how-many-of-each-class-of-library-layout-are-there">10.8.2. How many of each class of library layout are there?</a></li>
<li><a class="reference internal" href="#can-we-sort-the-file-by-library-layout-and-save-that-sorted-information-to-a-new-file">10.8.3. Can we sort the file by library layout and save that sorted information to a new file?</a></li>
<li><a class="reference internal" href="#can-we-extract-only-paired-end-records-into-a-new-file">10.8.4. Can we extract only paired-end records into a new file?</a></li>
</ul>
</li>
<li><a class="reference internal" href="#exercise">10.9. Exercise</a></li>
<li><a class="reference internal" href="#solution">10.10. Solution</a></li>
<li><a class="reference internal" href="#writing-for-loops">10.11. Writing for loops</a><ul>
<li><a class="reference internal" href="#using-basename-in-for-loops">10.11.1. Using Basename in for loops</a></li>
</ul>
</li>
</ul>
</li>
</ul>

    </div>
  </div>
</div>
  <h4>Previous topic</h4>
  <p class="topless"><a href="bash_lesson_1.html"
                        title="previous chapter">9. Introducing the Shell</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="bioconda_lesson.html"
                        title="next chapter">11. Working with Bioconda.</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/bash_lesson_2.md.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div class="sidebar-block">
  <div class="sidebar-wrapper">
    <div id="main-search">
      <form class="form-inline" action="search.html" method="GET" role="form">
        <div class="input-group">
          <input name="q" type="text" class="form-control" placeholder="Search...">
        </div>
        <input type="hidden" name="check_keywords" value="yes" />
        <input type="hidden" name="area" value="default" />
      </form>
    </div>
  </div>
</div>
      
    </div>
  </div>
        <div id="right-column">
          
          <div role="navigation" aria-label="breadcrumbs navigation">
            <ol class="breadcrumb">
              <li><a href="toc.html">Docs</a></li>
              
              <li>10. Bash lesson 2</li>
            </ol>
          </div>
          
          <div class="document clearer body">
            
  <div class="section" id="bash-lesson-2">
<h1>10. Bash lesson 2<a class="headerlink" href="#bash-lesson-2" title="Permalink to this headline">¶</a></h1>
<blockquote>
<div><p>Questions:</p>
<blockquote>
<div><ul class="simple">
<li>How can I view and search file contents?</li>
<li>How can I create, copy and delete files and directories?</li>
<li>How can I control who has permission to modify a file?</li>
<li>How can I repeat recently used commands?</li>
</ul>
</div></blockquote>
</div></blockquote>
<blockquote>
<div><p>Objectives:</p>
<blockquote>
<div><ul class="simple">
<li>View, search within, copy, move, and rename files. Create new directories.</li>
<li>Use wild cards (<code class="docutils literal notranslate"><span class="pre">*</span></code>) to perform operations on multiple files.</li>
<li>Make a file read only</li>
<li>Use the <code class="docutils literal notranslate"><span class="pre">history</span></code> command to view and repeat recently used commands.</li>
</ul>
</div></blockquote>
</div></blockquote>
<blockquote>
<div><p>Keypoints:</p>
<blockquote>
<div><ul class="simple">
<li>“You can view file contents using <code class="docutils literal notranslate"><span class="pre">less</span></code>, <code class="docutils literal notranslate"><span class="pre">cat</span></code>, <code class="docutils literal notranslate"><span class="pre">head</span></code> or <code class="docutils literal notranslate"><span class="pre">tail</span></code>.”</li>
<li>“The commands <code class="docutils literal notranslate"><span class="pre">cp</span></code>, <code class="docutils literal notranslate"><span class="pre">mv</span></code>, and <code class="docutils literal notranslate"><span class="pre">mkdir</span></code> are useful for manipulating existing files and creating new directories.”</li>
<li>“You can view file permissions using <code class="docutils literal notranslate"><span class="pre">ls</span> <span class="pre">-l</span></code> and change permissions using <code class="docutils literal notranslate"><span class="pre">chmod</span></code>.”</li>
<li>“The <code class="docutils literal notranslate"><span class="pre">history</span></code> command and the up arrow on your keyboard can be used to repeat recently used commands.”</li>
</ul>
</div></blockquote>
</div></blockquote>
<div class="section" id="working-with-files">
<h2>10.1. Working with Files<a class="headerlink" href="#working-with-files" title="Permalink to this headline">¶</a></h2>
<div class="section" id="our-data-set-fastq-files">
<h3>10.1.1. Our data set: FASTQ files<a class="headerlink" href="#our-data-set-fastq-files" title="Permalink to this headline">¶</a></h3>
<p>Now that we know how to navigate around our directory structure, lets start working with our sequencing files. We did a sequencing experiment and have two results files, which are stored in our <code class="docutils literal notranslate"><span class="pre">untrimmed_fastq</span></code> directory.</p>
</div>
<div class="section" id="wild-cards">
<h3>10.1.2. Wild cards<a class="headerlink" href="#wild-cards" title="Permalink to this headline">¶</a></h3>
<p>Navigate to your <code class="docutils literal notranslate"><span class="pre">untrimmed_fastq</span></code> directory.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd ~/dc_workshop/untrimmed_fastq
</pre></div>
</div>
<p>We are interested in looking at the FASTQ files in this directory. We can list all files with the .fastq extension using the command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ls *.fastq
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">SRR097977</span><span class="o">.</span><span class="n">fastq</span>  <span class="n">SRR098026</span><span class="o">.</span><span class="n">fastq</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">*</span></code> character is a special type of character called a wildcard, which can be used to represent any number of any type of character. Thus, <code class="docutils literal notranslate"><span class="pre">*.fastq</span></code> matches every file that ends with <code class="docutils literal notranslate"><span class="pre">.fastq</span></code>.</p>
<p>This command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ls *977.fastq
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">SRR097977</span><span class="o">.</span><span class="n">fastq</span>
</pre></div>
</div>
<p>lists only the file that ends with <code class="docutils literal notranslate"><span class="pre">977.fastq</span></code>.</p>
<p>We can use the command <code class="docutils literal notranslate"><span class="pre">echo</span></code> to see how the wildcard character is interpreted by the shell.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ echo *.fastq
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">SRR097977</span><span class="o">.</span><span class="n">fastq</span> <span class="n">SRR098026</span><span class="o">.</span><span class="n">fastq</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">*</span></code> is expanded to include any file that ends with <code class="docutils literal notranslate"><span class="pre">.fastq</span></code>.</p>
<p>This command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ls /usr/bin/*.sh
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">amuFormat</span><span class="o">.</span><span class="n">sh</span>  <span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">gettext</span><span class="o">.</span><span class="n">sh</span>  <span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">gvmap</span><span class="o">.</span><span class="n">sh</span>
</pre></div>
</div>
<p>Lists every file in <code class="docutils literal notranslate"><span class="pre">/usr/bin</span></code> that ends in the characters <code class="docutils literal notranslate"><span class="pre">.sh</span></code>.</p>
<blockquote>
<div><p><strong>Home vs. Root</strong></p>
<p>The <code class="docutils literal notranslate"><span class="pre">/</span></code> character is another navigational shortcut and refers to your root directory. The root directory is the highest level directory in your file system and contains files that are important for your computer to perform its daily work, but which you usually won’t
have to interact with directly. In our case, the root directory is two levels above our home directory, so <code class="docutils literal notranslate"><span class="pre">cd</span></code> or <code class="docutils literal notranslate"><span class="pre">cd</span> <span class="pre">~</span></code> will take you to <code class="docutils literal notranslate"><span class="pre">/home/dcuser</span></code> and <code class="docutils literal notranslate"><span class="pre">cd</span> <span class="pre">/</span></code> will take you to <code class="docutils literal notranslate"><span class="pre">/</span></code>, which is equivalent to <code class="docutils literal notranslate"><span class="pre">~/../../</span></code>. Try not to worry if this is confusing, it will all become clearer with practice.</p>
<p>While you will be using the root at the beginning of your absolute paths, it is important that you avoid working with data in these higher-level directories, as your commands can permanently alter files that the operating system needs to function. In many cases, trying to run commands in root directories will require special permissions which are not discussed here, so it’s best to avoid it and work within your home directory.</p>
</div></blockquote>
<blockquote>
<div><p><strong>Exercise</strong>
Do each of the following tasks from your current directory using a single
<code class="docutils literal notranslate"><span class="pre">ls</span></code> command for each.</p>
<ol class="simple">
<li>List all of the files in <code class="docutils literal notranslate"><span class="pre">/usr/bin</span></code> that start with the letter ‘c’.</li>
<li>List all of the files in <code class="docutils literal notranslate"><span class="pre">/usr/bin</span></code> that contain the letter ‘a’.</li>
<li>List all of the files in <code class="docutils literal notranslate"><span class="pre">/usr/bin</span></code> that end with the letter ‘o’.</li>
</ol>
<p><em>Bonus:</em> List all of the files in <code class="docutils literal notranslate"><span class="pre">/usr/bin</span></code> that contain the letter ‘a’ or the
letter ‘c’.</p>
<p><em>Hint:</em> The bonus question requires a Unix wildcard that we haven’t talked about yet. Trying searching the internet for information about Unix wildcards to find what you need to solve the bonus problem.</p>
<blockquote>
<div><p><em>Solution</em></p>
<ol class="simple">
<li><code class="docutils literal notranslate"><span class="pre">ls</span> <span class="pre">/usr/bin/c*</span></code></li>
<li><code class="docutils literal notranslate"><span class="pre">ls</span> <span class="pre">/usr/bin/*a*</span></code></li>
<li><code class="docutils literal notranslate"><span class="pre">ls</span> <span class="pre">/usr/bin/*o</span></code>
Bonus: <code class="docutils literal notranslate"><span class="pre">ls</span> <span class="pre">/usr/bin/*[ac]*</span></code></li>
</ol>
</div></blockquote>
</div></blockquote>
</div>
</div>
<div class="section" id="command-history">
<h2>10.2. Command History<a class="headerlink" href="#command-history" title="Permalink to this headline">¶</a></h2>
<p>If you want to repeat a command that you’ve run recently, you can access previous commands using the up arrow on your keyboard to go back to the most recent command. Likewise, the down arrow takes you forward in the command history.</p>
<p>A few more useful shortcuts:</p>
<ul class="simple">
<li><kbd>Ctrl</kbd> + <kbd>C</kbd> will cancel the command you are writing, and give you a
fresh prompt.</li>
<li><kbd>Ctrl</kbd>+<kbd>R</kbd> will do a reverse-search through your command history.  This
is very useful.</li>
<li><kbd>Ctrl</kbd>+<kbd>L</kbd> or the <code class="docutils literal notranslate"><span class="pre">clear</span></code> command will clear your screen.</li>
</ul>
<p>You can also review your recent commands with the <code class="docutils literal notranslate"><span class="pre">history</span></code> command, by entering:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ history
</pre></div>
</div>
<p>to see a numbered list of recent commands. You can reuse one of these commands directly by referring to the number of that command.</p>
<p>For example, if your history looked like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">259</span>  <span class="n">ls</span> <span class="o">*</span>
<span class="mi">260</span>  <span class="n">ls</span> <span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="nb">bin</span><span class="o">/*.</span><span class="n">sh</span>
<span class="mi">261</span>  <span class="n">ls</span> <span class="o">*</span><span class="n">R1</span><span class="o">*</span><span class="n">fastq</span>
</pre></div>
</div>
<p>then you could repeat command #260 by entering:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ !260
</pre></div>
</div>
<p>Type <code class="docutils literal notranslate"><span class="pre">!</span></code> (exclamation point) and then the number of the command from your history. You will be glad you learned this when you need to re-run very complicated commands.</p>
<blockquote>
<div><p><strong>Exercise</strong>
Find the line number in your history for the command that listed all the .sh
files in <code class="docutils literal notranslate"><span class="pre">/usr/bin</span></code>. Rerun that command.</p>
<blockquote>
<div><em>Solution</em>
First type <code class="docutils literal notranslate"><span class="pre">history</span></code>. Then use <code class="docutils literal notranslate"><span class="pre">!</span></code> followed by the line number to rerun that command.</div></blockquote>
</div></blockquote>
</div>
<div class="section" id="examining-files">
<h2>10.3. Examining Files<a class="headerlink" href="#examining-files" title="Permalink to this headline">¶</a></h2>
<p>We now know how to switch directories, run programs, and look at the contents of directories, but how do we look at the contents of files?</p>
<p>One way to examine a file is to print out all of the contents using the program <code class="docutils literal notranslate"><span class="pre">cat</span></code>.</p>
<p>Enter the following command from within the <code class="docutils literal notranslate"><span class="pre">untrimmed_fastq</span></code> directory:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cat SRR098026.fastq
</pre></div>
</div>
<p>This will print out all of the contents of the <code class="docutils literal notranslate"><span class="pre">SRR098026.fastq</span></code> to the screen.</p>
<blockquote>
<div><p><strong>Exercise</strong></p>
<ol class="simple">
<li>Print out the contents of the <code class="docutils literal notranslate"><span class="pre">~/dc_workshop/untrimmed_fastq/SRR097977.fastq</span></code> file. What is the last line of the file?</li>
<li>From your home directory, and without changing directories,
use one short command to print the contents of all of the files in
the <code class="docutils literal notranslate"><span class="pre">~/dc_workshop/untrimmed_fastq</span></code> directory.</li>
</ol>
<blockquote>
<div><p><em>Solution</em></p>
<ol class="simple">
<li>The last line of the file is <code class="docutils literal notranslate"><span class="pre">TC:CCC::CCCCCCCC&lt;8?6A:C28C&lt;608'&amp;&amp;&amp;,'$</span></code>.</li>
<li><code class="docutils literal notranslate"><span class="pre">cat</span> <span class="pre">~/dc_workshop/untrimmed_fastq/*</span></code></li>
</ol>
</div></blockquote>
</div></blockquote>
<p><code class="docutils literal notranslate"><span class="pre">cat</span></code> is a terrific program, but when the file is really big, it can be annoying to use. The program, <code class="docutils literal notranslate"><span class="pre">less</span></code>, is useful for this case. <code class="docutils literal notranslate"><span class="pre">less</span></code> opens the file as read only, and lets you navigate through it. The navigation commands are identical to the <code class="docutils literal notranslate"><span class="pre">man</span></code> program.</p>
<p>Enter the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ less SRR097977.fastq
</pre></div>
</div>
<p>Some navigation commands in <code class="docutils literal notranslate"><span class="pre">less</span></code></p>
<ul class="simple">
<li><kbd>Space</kbd> : to go forward</li>
<li><kbd>b</kbd>    : to go backward</li>
<li><kbd>g</kbd>    : to go to the beginning</li>
<li><kbd>G</kbd>    : to go to the end</li>
<li><kbd>q</kbd>    : to quit</li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">less</span></code> also gives you a way of searching through files. Use the “/” key to begin a search. Enter the word you would like to search for and press <code class="docutils literal notranslate"><span class="pre">enter</span></code>. The screen will jump to the next location where that word is found.</p>
<p>Remember, the <code class="docutils literal notranslate"><span class="pre">man</span></code> program actually uses <code class="docutils literal notranslate"><span class="pre">less</span></code> internally and therefore uses the same commands, so you can search documentation using “/” as well!</p>
<p>There’s another way that we can look at files, and in this case, just look at part of them. This can be particularly useful if we just want to see the beginning or end of the file, or see how it’s formatted.</p>
<p>The commands are <code class="docutils literal notranslate"><span class="pre">head</span></code> and <code class="docutils literal notranslate"><span class="pre">tail</span></code> and they let you look at the beginning and end of a file, respectively.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ head SRR098026.fastq
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>@SRR098026.1 HWUSI-EAS1599_1:2:1:0:968 length=35
NNNNNNNNNNNNNNNNCNNNNNNNNNNNNNNNNNN
+SRR098026.1 HWUSI-EAS1599_1:2:1:0:968 length=35
!!!!!!!!!!!!!!!!#!!!!!!!!!!!!!!!!!!
@SRR098026.2 HWUSI-EAS1599_1:2:1:0:312 length=35
NNNNNNNNNNNNNNNNANNNNNNNNNNNNNNNNNN
+SRR098026.2 HWUSI-EAS1599_1:2:1:0:312 length=35
!!!!!!!!!!!!!!!!#!!!!!!!!!!!!!!!!!!
@SRR098026.3 HWUSI-EAS1599_1:2:1:0:570 length=35
NNNNNNNNNNNNNNNNANNNNNNNNNNNNNNNNNN
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ tail SRR098026.fastq
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>+SRR098026.247 HWUSI-EAS1599_1:2:1:2:1311 length=35
#!##!#################!!!!!!!######
@SRR098026.248 HWUSI-EAS1599_1:2:1:2:118 length=35
GNTGNGGTCATCATACGCGCCCNNNNNNNGGCATG
+SRR098026.248 HWUSI-EAS1599_1:2:1:2:118 length=35
B!;?!A=5922:##########!!!!!!!######
@SRR098026.249 HWUSI-EAS1599_1:2:1:2:1057 length=35
CNCTNTATGCGTACGGCAGTGANNNNNNNGGAGAT
+SRR098026.249 HWUSI-EAS1599_1:2:1:2:1057 length=35
A!@B!BBB@ABAB#########!!!!!!!######
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">-n</span></code> option to either of these commands can be used to print the first or last <code class="docutils literal notranslate"><span class="pre">n</span></code> lines of a file.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ head -n 1 SRR098026.fastq
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nd">@SRR098026</span><span class="o">.</span><span class="mi">1</span> <span class="n">HWUSI</span><span class="o">-</span><span class="n">EAS1599_1</span><span class="p">:</span><span class="mi">2</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">0</span><span class="p">:</span><span class="mi">968</span> <span class="n">length</span><span class="o">=</span><span class="mi">35</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ tail -n 1 SRR098026.fastq
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>A!@B!BBB@ABAB#########!!!!!!!######
</pre></div>
</div>
</div>
<div class="section" id="details-on-the-fastq-format">
<h2>10.4. Details on the FASTQ format<a class="headerlink" href="#details-on-the-fastq-format" title="Permalink to this headline">¶</a></h2>
<p>Although it looks complicated (and it is), it’s easy to understand the <a class="reference external" href="https://en.wikipedia.org/wiki/FASTQ_format">fastq</a> format with a little decoding. Some rules about the format include…</p>
<ol class="simple">
<li>Always begins with ‘&#64;’ and then information about the read</li>
<li>The actual DNA sequence</li>
<li>Always begins with a ‘+’ and sometimes the same info in line 1</li>
<li>Has a string of characters which represent the quality scores; must have same number of characters as line 2</li>
</ol>
<p>We can view the first complete read in one of the files our dataset by using <code class="docutils literal notranslate"><span class="pre">head</span></code> to look at the first four lines.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ head -n4 SRR098026.fastq
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>@SRR098026.1 HWUSI-EAS1599_1:2:1:0:968 length=35
NNNNNNNNNNNNNNNNCNNNNNNNNNNNNNNNNNN
+SRR098026.1 HWUSI-EAS1599_1:2:1:0:968 length=35
!!!!!!!!!!!!!!!!#!!!!!!!!!!!!!!!!!!
</pre></div>
</div>
<p>All but one of the nucleotides in this read are unknown (<code class="docutils literal notranslate"><span class="pre">N</span></code>). This is a pretty bad read!</p>
<p>Line 4 shows the quality for each nucleotide in the read. Quality is interpreted as the probability of an incorrect base call (e.g. 1 in 10) or, equivalently, the base call accuracy (eg 90%). To make it possible to line up each individual nucleotide with its quality score, the numerical score is converted into a code where each individual character represents the numerical quality score for an individual nucleotide. For example, in the line above, the quality score line is:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>!!!!!!!!!!!!!!!!#!!!!!!!!!!!!!!!!!!
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">#</span></code> character and each of the <code class="docutils literal notranslate"><span class="pre">!</span></code> characters represent the encoded quality for an individual nucleotide. The numerical value assigned to each of these characters depends on the sequencing platform that generated the reads. The sequencing machine used to generate our data uses the standard Sanger quality PHRED score encoding, using by Illumina version 1.8 onwards. Each character is assigned a quality score between 0 and 40 as shown in the chart below.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Quality encoding: !&quot;#$%&amp;&#39;()*+,-./0123456789:;&lt;=&gt;?@ABCDEFGHI
                  |         |         |         |         |
Quality score:    0........10........20........30........40
</pre></div>
</div>
<p>Each quality score represents the probability that the corresponding nucleotide call is incorrect. This quality score is logarithmically based, so a quality score of 10 reflects a base call accuracy of 90%, but a quality score of 20 reflects a base call accuracy of 99%. These probability values are the results from the base calling algorithm and dependent on how much signal was captured for the base incorporation.</p>
<p>Looking back at our read:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>@SRR098026.1 HWUSI-EAS1599_1:2:1:0:968 length=35
NNNNNNNNNNNNNNNNCNNNNNNNNNNNNNNNNNN
+SRR098026.1 HWUSI-EAS1599_1:2:1:0:968 length=35
!!!!!!!!!!!!!!!!#!!!!!!!!!!!!!!!!!!
</pre></div>
</div>
<p>we can now see that the quality of each of the <code class="docutils literal notranslate"><span class="pre">N</span></code>s is 0 and the quality of the only nucleotide call (<code class="docutils literal notranslate"><span class="pre">C</span></code>) is also very poor (<code class="docutils literal notranslate"><span class="pre">#</span></code> = a quality score of 2). This is indeed a very bad read.</p>
</div>
<div class="section" id="creating-moving-copying-and-removing">
<h2>10.5. Creating, moving, copying, and removing<a class="headerlink" href="#creating-moving-copying-and-removing" title="Permalink to this headline">¶</a></h2>
<p>Now we can move around in the file structure, look at files, and search files. But what if we want to copy files or move them around or get rid of them? Most of the time, you can do these sorts of file manipulations without the command line, but there will be some cases (like when you’re working with a remote computer like we are for this lesson) where it will be impossible. You’ll also find that you may be working with hundreds of files and want to do similar manipulations to all of those files. In cases like this, it’s much faster to do these operations at the command line.</p>
<div class="section" id="copying-files">
<h3>10.5.1. Copying Files<a class="headerlink" href="#copying-files" title="Permalink to this headline">¶</a></h3>
<p>When working with computational data, it’s important to keep a safe copy of that data that can’t be accidentally overwritten or deleted. For this lesson, our raw data is our FASTQ files.  We don’t want to accidentally change the original files, so we’ll make a copy of them and change the file permissions so that we can read from, but not write to, the files.</p>
<p>First, let’s make a copy of one of our FASTQ files using the <code class="docutils literal notranslate"><span class="pre">cp</span></code> command.</p>
<p>Navigate to the <code class="docutils literal notranslate"><span class="pre">dc_workshop/untrimmed_fastq</span></code> directory and enter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cp SRR098026.fastq SRR098026-copy.fastq
$ ls -F
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">SRR097977</span><span class="o">.</span><span class="n">fastq</span>  <span class="n">SRR098026</span><span class="o">-</span><span class="n">copy</span><span class="o">.</span><span class="n">fastq</span>  <span class="n">SRR098026</span><span class="o">.</span><span class="n">fastq</span>
</pre></div>
</div>
<p>We now have two copies of the <code class="docutils literal notranslate"><span class="pre">SRR098026.fastq</span></code> file, one of them named <code class="docutils literal notranslate"><span class="pre">SRR098026-copy.fastq</span></code>. We’ll move this file to a new directory called <code class="docutils literal notranslate"><span class="pre">backup</span></code> where we’ll store our backup data files.</p>
</div>
<div class="section" id="creating-directories">
<h3>10.5.2. Creating Directories<a class="headerlink" href="#creating-directories" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">mkdir</span></code> command is used to make a directory. Enter <code class="docutils literal notranslate"><span class="pre">mkdir</span></code> followed by a space, then the directory name you want to create.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ mkdir backup
</pre></div>
</div>
</div>
<div class="section" id="moving-renaming">
<h3>10.5.3. Moving / Renaming<a class="headerlink" href="#moving-renaming" title="Permalink to this headline">¶</a></h3>
<p>We can now move our backup file to this directory. We can move files around using the command <code class="docutils literal notranslate"><span class="pre">mv</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ mv SRR098026-copy.fastq backup
$ ls backup
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">SRR098026</span><span class="o">-</span><span class="n">copy</span><span class="o">.</span><span class="n">fastq</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">mv</span></code> command is also how you rename files. Let’s rename this file to make it clear that this is a backup.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd backup
$ mv SRR098026-copy.fastq SRR098026-backup.fastq
$ ls
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">SRR098026</span><span class="o">-</span><span class="n">backup</span><span class="o">.</span><span class="n">fastq</span>
</pre></div>
</div>
</div>
<div class="section" id="file-permissions">
<h3>10.5.4. File Permissions<a class="headerlink" href="#file-permissions" title="Permalink to this headline">¶</a></h3>
<p>We’ve now made a backup copy of our file, but just because we have two copies doesn’t make us safe. We can still accidentally delete or overwrite both copies. To make sure we can’t accidentally mess up this backup file, we’re going to change the permissions on the file so that we’re only allowed to read (i.e. view) the file, not write to it (i.e. make new changes).</p>
<p>View the current permissions on a file using the <code class="docutils literal notranslate"><span class="pre">-l</span></code> (long) flag for the <code class="docutils literal notranslate"><span class="pre">ls</span></code> command.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ls -l
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">dcuser</span> <span class="n">dcuser</span> <span class="mi">43332</span> <span class="n">Nov</span> <span class="mi">15</span> <span class="mi">23</span><span class="p">:</span><span class="mi">02</span> <span class="n">SRR098026</span><span class="o">-</span><span class="n">backup</span><span class="o">.</span><span class="n">fastq</span>
</pre></div>
</div>
<p>The first part of the output for the <code class="docutils literal notranslate"><span class="pre">-l</span></code> flag gives you information about the file’s current permissions. There are ten slots in the permissions list. The first character in this list is related to file type, not permissions, so we’ll ignore it for now. The next three characters relate to the permissions that the file owner has, the next three relate to the permissions for group members, and the final three characters specify what other users outside of your group can do with the file. We’re going to concentrate on the three positions
that deal with your permissions (as the file owner).</p>
<p>Here the three positions that relate to the file owner are <code class="docutils literal notranslate"><span class="pre">rw-</span></code>. The <code class="docutils literal notranslate"><span class="pre">r</span></code> means that you have permission to read the file, the <code class="docutils literal notranslate"><span class="pre">w</span></code> indicates that you have permission to write to (i.e. make changes to) the file, and the third position is a <code class="docutils literal notranslate"><span class="pre">-</span></code>, indicating that you don’t have permission to carry out the ability encoded by that space (this is the space where <code class="docutils literal notranslate"><span class="pre">x</span></code> or executable ability is stored, we’ll talk more about this in <a class="reference external" href="http://www.datacarpentry.org/shell-genomics/05-writing-scripts/">a later lesson</a>).</p>
<p>Our goal for now is to change permissions on this file so that you no longer have <code class="docutils literal notranslate"><span class="pre">w</span></code> or write permissions. We can do this using the <code class="docutils literal notranslate"><span class="pre">chmod</span></code> (change mode) command and subtracting (<code class="docutils literal notranslate"><span class="pre">-</span></code>) the write permission <code class="docutils literal notranslate"><span class="pre">-w</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ chmod -w SRR098026-backup.fastq
$ ls -l
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">dcuser</span> <span class="n">dcuser</span> <span class="mi">43332</span> <span class="n">Nov</span> <span class="mi">15</span> <span class="mi">23</span><span class="p">:</span><span class="mi">02</span> <span class="n">SRR098026</span><span class="o">-</span><span class="n">backup</span><span class="o">.</span><span class="n">fastq</span>
</pre></div>
</div>
</div>
<div class="section" id="removing">
<h3>10.5.5. Removing<a class="headerlink" href="#removing" title="Permalink to this headline">¶</a></h3>
<p>To prove to ourselves that you no longer have the ability to modify this file, try deleting it with the <code class="docutils literal notranslate"><span class="pre">rm</span></code> command.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ rm SRR098026-backup.fastq
</pre></div>
</div>
<p>You’ll be asked if you want to override your file permissions.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>rm: remove write-protected regular file ‘SRR098026-backup.fastq’?
</pre></div>
</div>
<p>If you enter <code class="docutils literal notranslate"><span class="pre">n</span></code> (for no), the file will not be deleted. If you enter <code class="docutils literal notranslate"><span class="pre">y</span></code>, you will delete the file. This gives us an extra measure of security, as there is one more step between us and deleting our data files.</p>
<p><strong>Important</strong>: The <code class="docutils literal notranslate"><span class="pre">rm</span></code> command permanently removes the file. Be careful with this command. It doesn’t
just nicely put the files in the Trash. They’re really gone.</p>
<p>By default, <code class="docutils literal notranslate"><span class="pre">rm</span></code> will not delete directories. You can tell <code class="docutils literal notranslate"><span class="pre">rm</span></code> to delete a directory using the <code class="docutils literal notranslate"><span class="pre">-r</span></code> (recursive) option. Let’s delete the backup directory we just made.</p>
<p>Enter the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd ..
$ rm -r backup
</pre></div>
</div>
<p>This will delete not only the directory, but all files within the directory. If you have write-protected files in the directory, you will be asked whether you want to override your permission settings.</p>
<blockquote>
<div><p><strong>Exercise</strong></p>
<p>Starting in the <code class="docutils literal notranslate"><span class="pre">dc_workshop/untrimmed_fastq/</span></code> directory, do the following:</p>
<ol class="simple">
<li>Make sure that you have deleted your backup directory and all files it contains.</li>
<li>Create a copy of each of your FASTQ files. (Note: You’ll need to do this individually for each of the two FASTQ files. We haven’t
learned yet how to do this
with a wild-card.)</li>
<li>Use a wildcard to move all of your backup files to a new backup directory.</li>
<li>Change the permissions on all of your backup files to be write-protected.</li>
</ol>
<blockquote>
<div><p><em>Solution</em></p>
<ol class="simple">
<li><code class="docutils literal notranslate"><span class="pre">rm</span> <span class="pre">-r</span> <span class="pre">backup</span></code></li>
<li><code class="docutils literal notranslate"><span class="pre">cp</span> <span class="pre">SRR098026.fastq</span> <span class="pre">SRR098026-backup.fastq</span></code> and <code class="docutils literal notranslate"><span class="pre">cp</span> <span class="pre">SRR097977.fastq</span> <span class="pre">SRR097977-backup.fastq</span></code></li>
<li><code class="docutils literal notranslate"><span class="pre">mkdir</span> <span class="pre">backup</span></code> and <code class="docutils literal notranslate"><span class="pre">mv</span> <span class="pre">*-backup.fastq</span> <span class="pre">backup</span></code></li>
<li><code class="docutils literal notranslate"><span class="pre">chmod</span> <span class="pre">-w</span> <span class="pre">backup/*-backup.fastq</span></code>
It’s always a good idea to check your work with <code class="docutils literal notranslate"><span class="pre">ls</span> <span class="pre">-l</span> <span class="pre">backup</span></code>. You should see something like:</li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">dcuser</span> <span class="n">dcuser</span> <span class="mi">47552</span> <span class="n">Nov</span> <span class="mi">15</span> <span class="mi">23</span><span class="p">:</span><span class="mi">06</span> <span class="n">SRR097977</span><span class="o">-</span><span class="n">backup</span><span class="o">.</span><span class="n">fastq</span>
<span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">dcuser</span> <span class="n">dcuser</span> <span class="mi">43332</span> <span class="n">Nov</span> <span class="mi">15</span> <span class="mi">23</span><span class="p">:</span><span class="mi">06</span> <span class="n">SRR098026</span><span class="o">-</span><span class="n">backup</span><span class="o">.</span><span class="n">fastq</span>
</pre></div>
</div>
</div></blockquote>
</div></blockquote>
<hr class="docutils" />
<blockquote>
<div><p>Objectives:</p>
<ul class="simple">
<li>“Employ the <code class="docutils literal notranslate"><span class="pre">grep</span></code> command to search for information within files.”</li>
<li>“Print the results of a command to a file.”</li>
<li>“Construct command pipelines with two or more stages.”
Keypoints:</li>
<li>“<code class="docutils literal notranslate"><span class="pre">grep</span></code> is a powerful search tool with many options for customization.”</li>
<li>“<code class="docutils literal notranslate"><span class="pre">&gt;</span></code>, <code class="docutils literal notranslate"><span class="pre">&gt;&gt;</span></code>, and <code class="docutils literal notranslate"><span class="pre">|</span></code> are different ways of redirecting output.”</li>
<li>“<code class="docutils literal notranslate"><span class="pre">command</span> <span class="pre">&gt;</span> <span class="pre">file</span></code> redirects a command’s output to a file.”</li>
<li>“<code class="docutils literal notranslate"><span class="pre">command</span> <span class="pre">&gt;&gt;</span> <span class="pre">file</span></code> redirects a command’s output to a file without overwriting the existing contents of the file.”</li>
<li>“<code class="docutils literal notranslate"><span class="pre">command_1</span> <span class="pre">|</span> <span class="pre">command_2</span></code> redirects the output of the first command as input to the second command.”</li>
<li>“for loops are used for iteration”</li>
<li>“<code class="docutils literal notranslate"><span class="pre">basename</span></code> gets rid of repetitive parts of names”</li>
</ul>
</div></blockquote>
</div>
</div>
<hr class="docutils" />
<div class="section" id="searching-files">
<h2>10.6. Searching files<a class="headerlink" href="#searching-files" title="Permalink to this headline">¶</a></h2>
<p>We discussed in a previous episode how to search within a file using <code class="docutils literal notranslate"><span class="pre">less</span></code>. We can also search within files without even opening them, using <code class="docutils literal notranslate"><span class="pre">grep</span></code>. <code class="docutils literal notranslate"><span class="pre">grep</span></code> is a command-line utility for searching plain-text files for lines matching a specific set of characters (sometimes called a string) or a particular pattern (which can be specified using something called regular expressions). We’re not going to work with regular expressions in this lesson, and are instead going to specify the strings we are searching for.</p>
<p>Let’s give it a try!</p>
<blockquote>
<div><p><strong>Nucleotide abbreviations</strong></p>
<p>The four nucleotides that appear in DNA are abbreviated <code class="docutils literal notranslate"><span class="pre">A</span></code>, <code class="docutils literal notranslate"><span class="pre">C</span></code>, <code class="docutils literal notranslate"><span class="pre">T</span></code> and <code class="docutils literal notranslate"><span class="pre">G</span></code>. Unknown nucleotides are represented with the letter <code class="docutils literal notranslate"><span class="pre">N</span></code>. An <code class="docutils literal notranslate"><span class="pre">N</span></code> appearing
in a sequencing file represents a position where the sequencing machine was not able to confidently determine the nucleotide in that position. You can think of an <code class="docutils literal notranslate"><span class="pre">N</span></code> as a <code class="docutils literal notranslate"><span class="pre">NULL</span></code> value within a DNA sequence.</p>
</div></blockquote>
<p>We’ll search for strings inside of our fastq files. Let’s first make sure we are in the correct directory.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd ~/dc_workshop/untrimmed_fastq
</pre></div>
</div>
<p>Suppose we want to see how many reads in our file have really bad segments containing 10 consecutive unknown nucleoties (Ns). Let’s search for the string NNNNNNNNNN in the SRR098026 file.</p>
<blockquote>
<div><p><strong>Determining quality</strong></p>
<p>In this lesson, we’re going to be manually searching for strings of <code class="docutils literal notranslate"><span class="pre">N</span></code>s within our sequence results to illustrate some principles of file searching. It can be really useful to do this type of searching to get a feel for the quality of your sequencing results, however, in you research you will most likely use a bioinformatics tool that has a built-in program for filtering out low-quality reads. You’ll learn how to use one such tool in
<a class="reference external" href="http://www.datacarpentry.org/wrangling-genomics/00-readQC/">a later lesson</a>.</p>
</div></blockquote>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ grep NNNNNNNNNN SRR098026.fastq
</pre></div>
</div>
<p>This command returns a lot of output to the terminal. Every single line in the SRR098026 file that contains at least 10 consecutive Ns is printed to the terminal, regardless of how long or short the file is. We may be interested not only in the actual sequence which contains this string, but in the name (or identifier) of that sequence. We discussed in a previous lesson that the identifier line immediately precedes the nucleotide sequence for each read in a FASTQ file. We may also want to inspect the quality scores associated with each of these reads. To get all of this information, we will return the line immediately before each match and the two lines immediately after each match.</p>
<p>We can use the <code class="docutils literal notranslate"><span class="pre">-B</span></code> argument for grep to return a specific number of lines before each match and the <code class="docutils literal notranslate"><span class="pre">-A</span></code> argument to return a specific number of lines after each matching line. Here we want the line before and the two lines after each matching line so we add <code class="docutils literal notranslate"><span class="pre">-B1</span> <span class="pre">-A2</span></code> to our grep command.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ grep -B1 -A2 NNNNNNNNNN SRR098026.fastq
</pre></div>
</div>
<p>One of the sets of lines returned by this command is:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nd">@SRR098026</span><span class="o">.</span><span class="mi">177</span> <span class="n">HWUSI</span><span class="o">-</span><span class="n">EAS1599_1</span><span class="p">:</span><span class="mi">2</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">2025</span> <span class="n">length</span><span class="o">=</span><span class="mi">35</span>
<span class="n">CNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN</span>
<span class="o">+</span><span class="n">SRR098026</span><span class="o">.</span><span class="mi">177</span> <span class="n">HWUSI</span><span class="o">-</span><span class="n">EAS1599_1</span><span class="p">:</span><span class="mi">2</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">2025</span> <span class="n">length</span><span class="o">=</span><span class="mi">35</span>
<span class="c1">#!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!</span>
</pre></div>
</div>
<blockquote>
<div><p><strong>Exercise</strong></p>
<ol class="simple">
<li>Search for the sequence <code class="docutils literal notranslate"><span class="pre">GNATNACCACTTCC</span></code> in the <code class="docutils literal notranslate"><span class="pre">SRR098026.fastq</span></code> file.
Have your search return all matching lines and the name (or identifier) for each sequence
that contains a match.</li>
<li>Search for the sequence <code class="docutils literal notranslate"><span class="pre">AAGTT</span></code> in both FASTQ files.
Have your search return all matching lines and the name (or identifier) for each sequence
that contains a match.</li>
</ol>
<blockquote>
<div><p><em>Solution</em></p>
<ol class="simple">
<li><code class="docutils literal notranslate"><span class="pre">grep</span> <span class="pre">-B1</span> <span class="pre">GNATNACCACTTCC</span> <span class="pre">SRR098026.fastq</span></code></li>
<li><code class="docutils literal notranslate"><span class="pre">grep</span> <span class="pre">-B1</span> <span class="pre">AAGTT</span> <span class="pre">*.fastq</span></code></li>
</ol>
</div></blockquote>
</div></blockquote>
</div>
<div class="section" id="redirecting-output">
<h2>10.7. Redirecting output<a class="headerlink" href="#redirecting-output" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">grep</span></code> allowed us to identify sequences in our FASTQ files that match a particular pattern. All of these sequences were printed to our terminal screen, but in order to work with these sequences and perform other opperations on them, we will need to capture that output in some way.</p>
<p>We can do this with something called “redirection”. The idea is that we are taking what would ordinarily be printed to the terminal screen and redirecting it to another location. In our case, we want to print this information to a file so that we can look at it later and use other commands to analyze this data.</p>
<p>The command for redirecting output to a file is <code class="docutils literal notranslate"><span class="pre">&gt;</span></code>.</p>
<p>Let’s try out this command and copy all the records (including all four lines of each record) in our FASTQ files that contain ‘NNNNNNNNNN’ to another file called <code class="docutils literal notranslate"><span class="pre">bad_reads.txt</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ grep -B1 -A2 NNNNNNNNNN SRR098026.fastq &gt; bad_reads.txt
</pre></div>
</div>
<blockquote>
<div><p><strong>File extensions</strong></p>
<p>You might be confused about why we’re naming our output file with a <code class="docutils literal notranslate"><span class="pre">.txt</span></code> extension. After all,
it will be holding FASTQ formatted data that we’re extracting from our FASTQ files. Won’t it
also be a FASTQ file? The answer is, yes - it will be a FASTQ file and it would make sense to
name it with a <code class="docutils literal notranslate"><span class="pre">.fastq</span></code> extension. However, using a <code class="docutils literal notranslate"><span class="pre">.fastq</span></code> extension will lead us to problems
when we move to using wildcards later in this episode. We’ll point out where this becomes
important. For now, it’s good that you’re thinking about file extensions!</p>
</div></blockquote>
<p>The prompt should sit there a little bit, and then it should look like nothing happened. But type <code class="docutils literal notranslate"><span class="pre">ls</span></code>. You should see a new file called <code class="docutils literal notranslate"><span class="pre">bad_reads.txt</span></code>.</p>
<p>We can check the number of lines in our new file using a command called <code class="docutils literal notranslate"><span class="pre">wc</span></code>. <code class="docutils literal notranslate"><span class="pre">wc</span></code> stands for <strong>word count</strong>. This command counts the number of words, lines, and characters in a file.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ wc bad_reads.txt
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>  <span class="mi">537</span>  <span class="mi">1073</span> <span class="mi">23217</span> <span class="n">bad_reads</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>This will tell us the number of lines, words and characters in the file. If we want only the number of lines, we can use the <code class="docutils literal notranslate"><span class="pre">-l</span></code> flag for <code class="docutils literal notranslate"><span class="pre">lines</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ wc -l bad_reads.txt
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">537</span> <span class="n">bad_reads</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>Because we asked <code class="docutils literal notranslate"><span class="pre">grep</span></code> for all four lines of each FASTQ record, we need to divide the output by four to get the number of sequences that match our search pattern.</p>
<blockquote>
<div><p><strong>Exercise</strong></p>
<p>How many sequences in <code class="docutils literal notranslate"><span class="pre">SRR098026.fastq</span></code> contain at least 3 consecutive Ns?</p>
<blockquote>
<div><p><em>Solution</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ grep NNN SRR098026.fastq &gt; bad_reads.txt
$ wc -l bad_reads.txt
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">249</span>
</pre></div>
</div>
</div></blockquote>
</div></blockquote>
<p>We might want to search multiple FASTQ files for sequences that match our search pattern. However, we need to be careful, because each time we use the <code class="docutils literal notranslate"><span class="pre">&gt;</span></code> command to redirect output to a file, the new output will replace the output that was already present in the file. This is called “overwriting” and, just like you don’t want to overwrite your video recording of your kid’s first birthday party, you also want to avoid overwriting your data files.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ grep -B1 -A2 NNNNNNNNNN SRR098026.fastq &gt; bad_reads.txt
$ wc -l bad_reads.txt
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">537</span> <span class="n">bad_reads</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ grep -B1 -A2 NNNNNNNNNN SRR097977.fastq &gt; bad_reads.txt
$ wc -l bad_reads.txt
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">0</span> <span class="n">bad_reads</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>Here, the output of our second  call to <code class="docutils literal notranslate"><span class="pre">wc</span></code> shows that we no longer have any lines in our <code class="docutils literal notranslate"><span class="pre">bad_reads.txt</span></code> file. This is because the second file we searched (<code class="docutils literal notranslate"><span class="pre">SRR097977.fastq</span></code>) does not contain any lines that match our
search sequence. So our file was overwritten and is now empty.</p>
<p>We can avoid overwriting our files by using the command <code class="docutils literal notranslate"><span class="pre">&gt;&gt;</span></code>. <code class="docutils literal notranslate"><span class="pre">&gt;&gt;</span></code> is known as the “append redirect” and will
append new output to the end of a file, rather than overwriting it.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ grep -B1 -A2 NNNNNNNNNN SRR098026.fastq &gt; bad_reads.txt
$ wc -l bad_reads.txt
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">537</span> <span class="n">bad_reads</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ grep -B1 -A2 NNNNNNNNNN SRR097977.fastq &gt;&gt; bad_reads.txt
$ wc -l bad_reads.txt
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">537</span> <span class="n">bad_reads</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>The output of our second call to <code class="docutils literal notranslate"><span class="pre">wc</span></code> shows that we have not overwritten our original data.</p>
<p>We can also do this with a single line of code by using a wildcard.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ grep -B1 -A2 NNNNNNNNNN *.fastq &gt; bad_reads.txt
$ wc -l bad_reads.txt
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">537</span> <span class="n">bad_reads</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<blockquote>
<div><p><strong>File extensions - part 2</strong></p>
<p>This is where we would have trouble if we were naming our output file with a <code class="docutils literal notranslate"><span class="pre">.fastq</span></code> extension.
If we already had a file called <code class="docutils literal notranslate"><span class="pre">bad_reads.fastq</span></code> (from our previous <code class="docutils literal notranslate"><span class="pre">grep</span></code> practice)
and then ran the command above using a <code class="docutils literal notranslate"><span class="pre">.fastq</span></code> extension instead of a <code class="docutils literal notranslate"><span class="pre">.txt</span></code> extension, <code class="docutils literal notranslate"><span class="pre">grep</span></code>
would give us a warning.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">grep</span> <span class="o">-</span><span class="n">B1</span> <span class="o">-</span><span class="n">A2</span> <span class="n">NNNNNNNNNN</span> \<span class="o">*.</span><span class="n">fastq</span> <span class="o">&gt;</span> <span class="n">bad_reads</span><span class="o">.</span><span class="n">fastq</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>grep: input file ‘bad_reads.fastq’ is also the output
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">grep</span></code> is letting you know that the output file <code class="docutils literal notranslate"><span class="pre">bad_reads.fastq</span></code> is also included in your
<code class="docutils literal notranslate"><span class="pre">grep</span></code> call because it matches the <code class="docutils literal notranslate"><span class="pre">*.fastq</span></code> pattern. Be careful with this as it can lead to
some surprising output.</p>
</div></blockquote>
<p>Since we might have multiple different criteria we want to search for, creating a new output file each time has the potential to clutter up our workspace. We also so far haven’t been interested in the actual contents of those files, only in the number of reads that we’ve found. We created the files to store the reads and then counted the lines in the file to see how many reads matched our criteria. There’s a way to do this, however, that doesn’t require us to create these intermediate files - the pipe command (<code class="docutils literal notranslate"><span class="pre">|</span></code>).</p>
<p>This is probably not a key on your keyboard you use very much, so let’s all take a minute to find that key.
What <code class="docutils literal notranslate"><span class="pre">|</span></code> does is take the output that is scrolling by on the terminal and uses that output as input to another command. When our output was scrolling by, we might have wished we could slow it down and look at it, like we can with <code class="docutils literal notranslate"><span class="pre">less</span></code>. Well it turns out that we can! We can redirect our output from our <code class="docutils literal notranslate"><span class="pre">grep</span></code> call through the <code class="docutils literal notranslate"><span class="pre">less</span></code> command.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ grep -B1 -A2 NNNNNNNNNN SRR098026.fastq | less
</pre></div>
</div>
<p>We can now see the output from our <code class="docutils literal notranslate"><span class="pre">grep</span></code> call within the <code class="docutils literal notranslate"><span class="pre">less</span></code> interface. We can use the up and down arrows
to scroll through the output and use <code class="docutils literal notranslate"><span class="pre">q</span></code> to exit <code class="docutils literal notranslate"><span class="pre">less</span></code>.</p>
<p>Redirecting output is often not intuitive, and can take some time to get used to. Once you’re comfortable with redirection, however, you’ll be able to combine any number of commands to do all sorts of exciting things with your data!</p>
<p>None of the command line programs we’ve been learning do anything all that impressive on their own, but when you start chaining them together, you can do some really powerful things very efficiently. Let’s take a few minutes to practice.</p>
<blockquote>
<div><p><strong>Exercise</strong></p>
<p>Now that we know about the pipe (<code class="docutils literal notranslate"><span class="pre">|</span></code>), write a single command to find the number of reads
in the <code class="docutils literal notranslate"><span class="pre">SRR098026.fastq</span></code> file that contain at least two regions of 5 unknown
nucleotides in a row, separated by any number of known nucleotides. Do this without creating
a new file.</p>
<blockquote>
<div><p><em>Solution</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ grep &quot;NNNNN*NNNNN&quot; SRR098026.fastq | wc -l
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">186</span>
</pre></div>
</div>
</div></blockquote>
</div></blockquote>
</div>
<div class="section" id="file-manipulation-and-more-practice-with-pipes">
<h2>10.8. File manipulation and more practice with pipes<a class="headerlink" href="#file-manipulation-and-more-practice-with-pipes" title="Permalink to this headline">¶</a></h2>
<p>Let’s use the tools we’ve added to our tool kit so far, along with a few new ones, to example our SRA metadata file. First, let’s navigate to the correct directory.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd
$ cd dc_workshop/sra_metadata
</pre></div>
</div>
<p>This file contains a lot of information about the samples that we submitted for sequencing. We took a look at this file in an earlier lesson. Here we’re going to use the information in this file to answer some questions about our samples.</p>
<div class="section" id="how-many-of-the-read-libraries-are-paired-end">
<h3>10.8.1. How many of the read libraries are paired end?<a class="headerlink" href="#how-many-of-the-read-libraries-are-paired-end" title="Permalink to this headline">¶</a></h3>
<p>The samples that we submitted to the sequencing facility were a mix of single and paired end libraries. We know that we recorded information in our metadata table about which samples used which library preparation method, but we don’t remember exactly where this data is recorded. Let’s start by looking at our column headers to see which column might have this information. Our column headers are in the first row of our data table, so we can use <code class="docutils literal notranslate"><span class="pre">head</span></code> with a <code class="docutils literal notranslate"><span class="pre">-n</span></code> flag to look at just the first row of the file.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ head -n 1 SraRunTable.txt
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">BioSample_s</span>	<span class="n">InsertSize_l</span>	<span class="n">LibraryLayout_s</span>	<span class="n">Library_Name_s</span>	<span class="n">LoadDate_s</span>	<span class="n">MBases_l</span>	<span class="n">MBytes_l</span>	<span class="n">ReleaseDate_s</span> <span class="n">Run_s</span> <span class="n">SRA_Sample_s</span> <span class="n">Sample_Name_s</span> <span class="n">Assay_Type_s</span> <span class="n">AssemblyName_s</span> <span class="n">BioProject_s</span> <span class="n">Center_Name_s</span> <span class="n">Consent_s</span> <span class="n">Organism_Platform_s</span> <span class="n">SRA_Study_s</span> <span class="n">g1k_analysis_group_s</span> <span class="n">g1k_pop_code_s</span> <span class="n">source_s</span> <span class="n">strain_s</span>
</pre></div>
</div>
<p>That is only the first line of our file, but because there are a lot of columns, the output likely wraps around your terminal window and appears as multiple lines. Once we figure out which column our data is in, we can use a command called <code class="docutils literal notranslate"><span class="pre">cut</span></code> to extract the column of interest.</p>
<p>Because this is pretty hard to read, we can look at just a few column header names at a time by combining the <code class="docutils literal notranslate"><span class="pre">|</span></code> redirect and <code class="docutils literal notranslate"><span class="pre">cut</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ head -n 1 SraRunTable.txt | cut -f1-4
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">cut</span></code> takes a <code class="docutils literal notranslate"><span class="pre">-f</span></code> flag, which stands for “field”. This flag accepts a list of field numbers, in our case, column numbers. Here we are extracting the first four column names.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">BioSample_s</span> <span class="n">InsertSize_l</span>      <span class="n">LibraryLayout_s</span>	<span class="n">Library_Name_s</span>
</pre></div>
</div>
<p>The Library Layout_s column looks like it should have the information we want.  Let’s look at some of the data from that column. We can use <code class="docutils literal notranslate"><span class="pre">cut</span></code> to extract only the 3rd column from the file and then use the <code class="docutils literal notranslate"><span class="pre">|</span></code> operator with <code class="docutils literal notranslate"><span class="pre">head</span></code> to look at just the first few lines of data in that column.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cut -f3 SraRunTable.txt | head -n 10
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Library</span> <span class="n">Layout_s</span>
<span class="n">SINGLE</span>
<span class="n">SINGLE</span>
<span class="n">SINGLE</span>
<span class="n">SINGLE</span>
<span class="n">SINGLE</span>
<span class="n">SINGLE</span>
<span class="n">SINGLE</span>
<span class="n">SINGLE</span>
<span class="n">PAIRED</span>
</pre></div>
</div>
<p>We can see that there are (at least) two categories, SINGLE and PAIRED.  We want to search all entries in this column for just PAIRED and count the number of matches. For this, we will use the <code class="docutils literal notranslate"><span class="pre">|</span></code> operator twice
to combine <code class="docutils literal notranslate"><span class="pre">cut</span></code> (to extract the column we want), <code class="docutils literal notranslate"><span class="pre">grep</span></code> (to find matches) and <code class="docutils literal notranslate"><span class="pre">wc</span></code> (to count matches).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cut -f3 SraRunTable.txt | grep PAIRED | wc -l
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">2</span>
</pre></div>
</div>
<p>We can see from this that we have only two paired-end libraries in the samples we submitted for
sequencing.</p>
<blockquote>
<div><p><strong>Exercise</strong></p>
<p>How many single-end libraries are in our samples?</p>
<blockquote>
<div><p><em>Solution</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cut -f3 SraRunTable.txt | grep SINGLE | wc -l
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">35</span>
</pre></div>
</div>
</div></blockquote>
</div></blockquote>
</div>
<div class="section" id="how-many-of-each-class-of-library-layout-are-there">
<h3>10.8.2. How many of each class of library layout are there?<a class="headerlink" href="#how-many-of-each-class-of-library-layout-are-there" title="Permalink to this headline">¶</a></h3>
<p>We can extract even more information from our metadata table if we add in some new tools: <code class="docutils literal notranslate"><span class="pre">sort</span></code> and <code class="docutils literal notranslate"><span class="pre">uniq</span></code>. The <code class="docutils literal notranslate"><span class="pre">sort</span></code> command will sort the lines of a text file and the <code class="docutils literal notranslate"><span class="pre">uniq</span></code> command will filter out repeated neighboring lines in a file. You might expect <code class="docutils literal notranslate"><span class="pre">uniq</span></code> to extract all of the unique lines in a file. This isn’t what it does, however, for reasons involving computer memory and speed. If we want to extract all unique lines, we can do so by combining <code class="docutils literal notranslate"><span class="pre">uniq</span></code> with <code class="docutils literal notranslate"><span class="pre">sort</span></code>. We’ll see how to do this soon.</p>
<p>For example, if we want to know how many samples of each library type are recorded in our table, we can extract the third column (with <code class="docutils literal notranslate"><span class="pre">cut</span></code>), and pipe that output into <code class="docutils literal notranslate"><span class="pre">sort</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cut -f3 SraRunTable.txt | sort
</pre></div>
</div>
<p>If you look closely, you might see that we have one line that reads “Library Layout_s”. This is the header of our column. We can discard this information using the <code class="docutils literal notranslate"><span class="pre">-v</span></code> flag in <code class="docutils literal notranslate"><span class="pre">grep</span></code>, which means return all the lines that <strong>do not</strong> match the search pattern.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cut -f3 SraRunTable.txt | grep -v LibraryLayout_s | sort
</pre></div>
</div>
<p>This command returns a sorted list (too long to show here) of PAIRED and SINGLE values. We can use the <code class="docutils literal notranslate"><span class="pre">uniq</span></code> command to see a list of all the different categories that are present. If we do this, we see that the only two types of libraries we have present are labelled PAIRED and SINGLE. There aren’t any other types in our file.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cut -f3 SraRunTable.txt | grep -v LibraryLayout_s | sort | uniq
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">PAIRED</span>
<span class="n">SINGLE</span>
</pre></div>
</div>
<p>If we want to count how many of each we have, we can use the <code class="docutils literal notranslate"><span class="pre">-c</span></code> (count) flag for <code class="docutils literal notranslate"><span class="pre">uniq</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cut -f3 SraRunTable.txt | grep -v LibraryLayout_s | sort | uniq -c
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">2</span> <span class="n">PAIRED</span>
<span class="mi">35</span> <span class="n">SINGLE</span>
</pre></div>
</div>
<blockquote>
<div><p><strong>Exercise</strong></p>
<ol class="simple">
<li>How many different sample load dates are there?</li>
<li>How many samples were loaded on each date?</li>
</ol>
<blockquote>
<div><p><em>Solution</em></p>
<ol>
<li><p class="first">There are two different sample load dates.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cut</span> <span class="o">-</span><span class="n">f5</span> <span class="n">SraRunTable</span><span class="o">.</span><span class="n">txt</span> <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">v</span> <span class="n">LoadDate_s</span> <span class="o">|</span> <span class="n">sort</span> <span class="o">|</span> <span class="n">uniq</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">25</span><span class="o">-</span><span class="n">Jul</span><span class="o">-</span><span class="mi">12</span>
<span class="mi">29</span><span class="o">-</span><span class="n">May</span><span class="o">-</span><span class="mi">14</span>
</pre></div>
</div>
</li>
<li><p class="first">Six samples were loaded on one date and 31 were loaded on the other.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cut</span> <span class="o">-</span><span class="n">f5</span> <span class="n">SraRunTable</span><span class="o">.</span><span class="n">txt</span> <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">v</span> <span class="n">LoadDate_s</span> <span class="o">|</span> <span class="n">sort</span> <span class="o">|</span> <span class="n">uniq</span> <span class="o">-</span><span class="n">c</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span> <span class="mi">6</span> <span class="mi">25</span><span class="o">-</span><span class="n">Jul</span><span class="o">-</span><span class="mi">12</span>
<span class="mi">31</span> <span class="mi">29</span><span class="o">-</span><span class="n">May</span><span class="o">-</span><span class="mi">14</span>
</pre></div>
</div>
</li>
</ol>
</div></blockquote>
</div></blockquote>
</div>
<div class="section" id="can-we-sort-the-file-by-library-layout-and-save-that-sorted-information-to-a-new-file">
<h3>10.8.3. Can we sort the file by library layout and save that sorted information to a new file?<a class="headerlink" href="#can-we-sort-the-file-by-library-layout-and-save-that-sorted-information-to-a-new-file" title="Permalink to this headline">¶</a></h3>
<p>We might want to re-order our entire metadata table so that all of the paired-end samples appear together and all of the single-end samples appear together. We can use the <code class="docutils literal notranslate"><span class="pre">-k</span></code> (key) flag for <code class="docutils literal notranslate"><span class="pre">sort</span></code> to sort based on a particular column. This is similar to the <code class="docutils literal notranslate"><span class="pre">-f</span></code> flag for <code class="docutils literal notranslate"><span class="pre">cut</span></code>.</p>
<p>Let’s sort based on the third column (<code class="docutils literal notranslate"><span class="pre">-k3</span></code>) and redirect our output to a new file.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ sort -k3 SraRunTable.txt &gt; SraRunTable_sorted_by_layout.txt
</pre></div>
</div>
</div>
<div class="section" id="can-we-extract-only-paired-end-records-into-a-new-file">
<h3>10.8.4. Can we extract only paired-end records into a new file?<a class="headerlink" href="#can-we-extract-only-paired-end-records-into-a-new-file" title="Permalink to this headline">¶</a></h3>
<p>We also might want to extract the information for all samples that meet a specific criterion (for example, are paired-end) and put those lines of our table in a new file. First, we need to check to make sure that the pattern we’re searching for (“PAIRED”) only appears in the column where we expect it to occur (column 3). We know from earlier that there are only two paired-end samples in the file, so we can <code class="docutils literal notranslate"><span class="pre">grep</span></code> for “PAIRED” and see how many results we get.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ grep PAIRED SraRunTable.txt | wc -l
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">2</span>
</pre></div>
</div>
<p>There are only two results, so we can use “PAIRED” as our search term to extract the paired-end samples to a new file.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ grep PAIRED SraRunTable.txt &gt; SraRunTable_only_paired_end.txt
</pre></div>
</div>
</div>
</div>
<div class="section" id="exercise">
<h2>10.9. Exercise<a class="headerlink" href="#exercise" title="Permalink to this headline">¶</a></h2>
<p>Sort samples by load date and export each of those sets to a new file (one new file per
unique load date).</p>
</div>
<div class="section" id="solution">
<h2>10.10. Solution<a class="headerlink" href="#solution" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ grep 25-Jul-12 SraRunTable.txt &gt; SraRunTable_25-Jul-12.txt
$ grep 29-May-14 SraRunTable.txt &gt; SraRunTable_29-May-14.txt
</pre></div>
</div>
</div>
<div class="section" id="writing-for-loops">
<h2>10.11. Writing for loops<a class="headerlink" href="#writing-for-loops" title="Permalink to this headline">¶</a></h2>
<p>Loops are key to productivity improvements through automation as they allow us to execute commands repeatedly.
Similar to wildcards and tab completion, using loops also reduces the amount of typing (and typing mistakes).
Loops are helpful when performing operations on groups of sequencing files, such as unzipping or trimming multiple files. We will use loops for these purposes in subsequent analyses, but will cover the basics of them for now.</p>
<p>When the shell sees the keyword <code class="docutils literal notranslate"><span class="pre">for</span></code>, it knows to repeat a command (or group of commands) once for each item in a list. Each time the loop runs (called an iteration), an item in the list is assigned in sequence to the <strong>variable</strong>, and the commands inside the loop are executed, before moving on to  the next item in the list. Inside the loop, we call for the variable’s value by putting <code class="docutils literal notranslate"><span class="pre">$</span></code> in front of it. The <code class="docutils literal notranslate"><span class="pre">$</span></code> tells the shell interpreter to treat the <strong>variable</strong> as a variable name and substitute its value in its place, rather than treat it as text or an external command.</p>
<p>Let’s write a for loop to show us the first two lines of the fastq files we downloaded earlier. You will notice shell prompt changes from <code class="docutils literal notranslate"><span class="pre">$</span></code> to <code class="docutils literal notranslate"><span class="pre">&gt;</span></code> and back again as we were typing in our loop. The second prompt, <code class="docutils literal notranslate"><span class="pre">&gt;</span></code>, is different to remind us that we haven’t finished typing a complete command yet. A semicolon, <code class="docutils literal notranslate"><span class="pre">;</span></code>, can be used to separate two commands written on a single line.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd ../untrimmed_fastq/
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ for filename in *.fastq
&gt; do
&gt; head -n 2 ${filename}
&gt; done
</pre></div>
</div>
<p>The for loop begins with the formula <code class="docutils literal notranslate"><span class="pre">for</span> <span class="pre">&lt;variable&gt;</span> <span class="pre">in</span> <span class="pre">&lt;group</span> <span class="pre">to</span> <span class="pre">iterate</span> <span class="pre">over&gt;</span></code>. In this case, the word <code class="docutils literal notranslate"><span class="pre">filename</span></code> is designated as the variable to be used over each iteration. In our case <code class="docutils literal notranslate"><span class="pre">SRR097977.fastq</span></code> and <code class="docutils literal notranslate"><span class="pre">SRR098026.fastq</span></code> will be substituted for <code class="docutils literal notranslate"><span class="pre">filename</span></code> because they fit the pattern of ending with .fastq in directory we’ve specified. The next line of the for loop is <code class="docutils literal notranslate"><span class="pre">do</span></code>. The next line is the code that we want to excute. We are telling the loop to print the first two lines of each variable we iterate over. Finally, the
word <code class="docutils literal notranslate"><span class="pre">done</span></code> ends the loop.</p>
<p>After executing the loop, you should see the first two lines of both fastq files printed to the terminal. Let’s create a loop that will save this information to a file.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ for filename in *.fastq
&gt; do
&gt; head -n 2 ${filename} &gt;&gt; seq_info.txt
&gt; done
</pre></div>
</div>
<p>Note that we are using <code class="docutils literal notranslate"><span class="pre">&gt;&gt;</span></code> to append the text to our <code class="docutils literal notranslate"><span class="pre">seq_info.txt</span></code> file. If we used <code class="docutils literal notranslate"><span class="pre">&gt;</span></code>, the <code class="docutils literal notranslate"><span class="pre">seq_info.txt</span></code> file would be rewritten every time the loop iterates, so it would only have text from the last variable used. Instead, <code class="docutils literal notranslate"><span class="pre">&gt;&gt;</span></code> adds to the end of the file.</p>
<div class="section" id="using-basename-in-for-loops">
<h3>10.11.1. Using Basename in for loops<a class="headerlink" href="#using-basename-in-for-loops" title="Permalink to this headline">¶</a></h3>
<p>Basename is a function in UNIX that is helpful for removing a uniform part of a name from a list of files. In this case, we will use basename to remove the <code class="docutils literal notranslate"><span class="pre">*.fastq</span></code> from the files that we’ve been working with. Inside our for loop, we create a new <code class="docutils literal notranslate"><span class="pre">name</span></code> variable. We call the <code class="docutils literal notranslate"><span class="pre">basename</span></code> function inside the parenthesis, then give our variable name from the for loop, in this case <code class="docutils literal notranslate"><span class="pre">$filename</span></code>, and finally state that <code class="docutils literal notranslate"><span class="pre">.fastq</span></code> should be removed from the file name. It’s important to note that we’re not changing the actual files, we’re creating and manipulating a new variable called <code class="docutils literal notranslate"><span class="pre">name</span></code>. The line <code class="docutils literal notranslate"><span class="pre">&gt;</span> <span class="pre">echo</span> <span class="pre">$name</span></code> will print to the terminal the variable <code class="docutils literal notranslate"><span class="pre">name</span></code>
each time the for loop runs. Because we are iterating over two files, we expect to see two lines of output.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ for filename in *.fastq
&gt; do
&gt; name=$(basename ${filename} .fastq)
&gt; echo ${name}
&gt; done
</pre></div>
</div>
<p>Although the utility of basename may still seem unclear, it will become very useful in subsequent analysis, such as trimming many reads in a for loop.</p>
</div>
</div>
</div>


          </div>
            
  <div class="footer-relations">
    
      <div class="pull-left">
        <a class="btn btn-default" href="bash_lesson_1.html" title="previous chapter (use the left arrow)">9. Introducing the Shell</a>
      </div>
    
      <div class="pull-right">
        <a class="btn btn-default" href="bioconda_lesson.html" title="next chapter (use the right arrow)">11. Working with Bioconda.</a>
      </div>
    </div>
    <div class="clearer"></div>
  
        </div>
        <div class="clearfix"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="bioconda_lesson.html" title="11. Working with Bioconda."
             >next</a> |</li>
        <li class="right" >
          <a href="bash_lesson_1.html" title="9. Introducing the Shell"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="toc.html">2019-04-11-gwu 0.0.1 documentation</a> &#187;</li> 
      </ul>
    </div>
<script type="text/javascript">
  $("#mobile-toggle a").click(function () {
    $("#left-column").toggle();
  });
</script>
<script type="text/javascript" src="_static/js/bootstrap.js"></script>
  <div class="footer">
    &copy; Copyright 2019, Sateesh Peri. Created using <a href="http://sphinx.pocoo.org/">Sphinx</a>.
  </div>
  </body>
</html>