
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  
  <!-- Licensed under the Apache 2.0 License -->
  <link rel="stylesheet" type="text/css" href="_static/fonts/open-sans/stylesheet.css" />
  <!-- Licensed under the SIL Open Font License -->
  <link rel="stylesheet" type="text/css" href="_static/fonts/source-serif-pro/source-serif-pro.css" />
  <link rel="stylesheet" type="text/css" href="_static/css/bootstrap.min.css" />
  <link rel="stylesheet" type="text/css" href="_static/css/bootstrap-theme.min.css" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
    <title>14. Assessing Read Quality &#8212; 2019-04-11-gwu 0.0.1 documentation</title>
    <link rel="stylesheet" href="_static/guzzle.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="17. Trimming and Filtering" href="trimming.html" />
    <link rel="prev" title="12. Workflow_Overview" href="workflow_overview.html" />
  
   

  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="trimming.html" title="17. Trimming and Filtering"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="workflow_overview.html" title="12. Workflow_Overview"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="toc.html">2019-04-11-gwu 0.0.1 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="container-wrapper">

      <div id="mobile-toggle">
        <a href="#"><span class="glyphicon glyphicon-align-justify" aria-hidden="true"></span></a>
      </div>
  <div id="left-column">
    <div class="sphinxsidebar">
<div class="sidebar-block">
  <div class="sidebar-wrapper">
    <h2>Contents</h2>
    <div class="sidebar-localtoc">
      <ul>
<li><a class="reference internal" href="#">14. Assessing Read Quality</a></li>
<li><a class="reference internal" href="#starting-with-data">15. Starting with Data</a></li>
<li><a class="reference internal" href="#quality-control">16. Quality Control</a><ul>
<li><a class="reference internal" href="#details-on-the-fastq-format">16.1. Details on the FASTQ format</a></li>
<li><a class="reference internal" href="#assessing-quality-using-fastqc">16.2. Assessing Quality using FastQC</a></li>
<li><a class="reference internal" href="#running-fastqc">16.3. Running FastQC</a><ul>
<li><a class="reference internal" href="#multiqc">16.3.1. MultiQC</a></li>
</ul>
</li>
<li><a class="reference internal" href="#viewing-the-fastqc-results">16.4. Viewing the FastQC results</a></li>
<li><a class="reference internal" href="#decoding-the-other-fastqc-outputs">16.5. Decoding the other FastQC outputs</a></li>
</ul>
</li>
</ul>

    </div>
  </div>
</div>
  <h4>Previous topic</h4>
  <p class="topless"><a href="workflow_overview.html"
                        title="previous chapter">12. Workflow_Overview</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="trimming.html"
                        title="next chapter">17. Trimming and Filtering</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/quality_control.md.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div class="sidebar-block">
  <div class="sidebar-wrapper">
    <div id="main-search">
      <form class="form-inline" action="search.html" method="GET" role="form">
        <div class="input-group">
          <input name="q" type="text" class="form-control" placeholder="Search...">
        </div>
        <input type="hidden" name="check_keywords" value="yes" />
        <input type="hidden" name="area" value="default" />
      </form>
    </div>
  </div>
</div>
      
    </div>
  </div>
        <div id="right-column">
          
          <div role="navigation" aria-label="breadcrumbs navigation">
            <ol class="breadcrumb">
              <li><a href="toc.html">Docs</a></li>
              
              <li>14. Assessing Read Quality</li>
            </ol>
          </div>
          
          <div class="document clearer body">
            
  <div class="section" id="assessing-read-quality">
<h1>14. Assessing Read Quality<a class="headerlink" href="#assessing-read-quality" title="Permalink to this headline">¶</a></h1>
<blockquote>
<div><p>Questions:</p>
<blockquote>
<div><ul class="simple">
<li>How can I describe the quality of my data?</li>
</ul>
</div></blockquote>
</div></blockquote>
<blockquote>
<div><p>Objectives:</p>
<blockquote>
<div><ul class="simple">
<li>“Explain how a FASTQ file encodes per-base quality scores.”</li>
<li>“Interpret a FastQC plot summarizing per-base quality across all reads.”</li>
<li>“Use <code class="docutils literal notranslate"><span class="pre">for</span></code> loops to automate operations on multiple files.”</li>
</ul>
</div></blockquote>
</div></blockquote>
<blockquote>
<div><p>Keypoints:</p>
<blockquote>
<div><ul class="simple">
<li>“Quality encodings vary across sequencing platforms.”</li>
<li>“<code class="docutils literal notranslate"><span class="pre">for</span></code> loops let you perform the same set of operations on multiple files with a single command.”</li>
</ul>
</div></blockquote>
</div></blockquote>
</div>
<hr class="docutils" />
<div class="section" id="starting-with-data">
<h1>15. Starting with Data<a class="headerlink" href="#starting-with-data" title="Permalink to this headline">¶</a></h1>
<blockquote>
<div>Often times, the first step in a bioinformatics workflow is getting the data you want to work with onto a computer where you can work with it.
If you have sequenced your own data, the sequencing center will usually provide you with a link that you can use to download your data. Today we will be working with publicly available sequencing data.</div></blockquote>
<blockquote>
<div><strong>About the Dataset:</strong> We are studying a population of <em>Escherichia coli</em> (designated Ara-3), which were propagated for more than 50,000 generations in a glucose-limited minimal medium. We will be working with three samples from this experiment, one from 5,000 generations, one from 15,000 generations, and one from 50,000 generations. The population changed substantially during the course of the experiment, and we will be exploring how with our variant calling workflow.
The data are paired-end, so we will download two files for each sample. We will use the <a class="reference external" href="https://www.ebi.ac.uk/ena">European Nucleotide Archive</a> to get our data. The ENA provides a comprehensive record of the world’s nucleotide sequencing information, covering raw sequencing data, sequence assembly information and functional annotation.</div></blockquote>
<p>First, create the project directory and sub-directories:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="o">~/</span>
<span class="n">mkdir</span> <span class="o">-</span><span class="n">p</span> <span class="n">dc_workshop</span><span class="o">/</span><span class="n">data</span>
<span class="n">cd</span> <span class="n">dc_workshop</span><span class="o">/</span><span class="n">data</span>
</pre></div>
</div>
<p>To download the data, run the commands below. It will take about 10 minutes to download the files.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">curl</span> <span class="o">-</span><span class="n">O</span> <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">sra</span><span class="o">.</span><span class="n">ebi</span><span class="o">.</span><span class="n">ac</span><span class="o">.</span><span class="n">uk</span><span class="o">/</span><span class="n">vol1</span><span class="o">/</span><span class="n">fastq</span><span class="o">/</span><span class="n">SRR258</span><span class="o">/</span><span class="mi">004</span><span class="o">/</span><span class="n">SRR2589044</span><span class="o">/</span><span class="n">SRR2589044_1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">curl</span> <span class="o">-</span><span class="n">O</span> <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">sra</span><span class="o">.</span><span class="n">ebi</span><span class="o">.</span><span class="n">ac</span><span class="o">.</span><span class="n">uk</span><span class="o">/</span><span class="n">vol1</span><span class="o">/</span><span class="n">fastq</span><span class="o">/</span><span class="n">SRR258</span><span class="o">/</span><span class="mi">004</span><span class="o">/</span><span class="n">SRR2589044</span><span class="o">/</span><span class="n">SRR2589044_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">curl</span> <span class="o">-</span><span class="n">O</span> <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">sra</span><span class="o">.</span><span class="n">ebi</span><span class="o">.</span><span class="n">ac</span><span class="o">.</span><span class="n">uk</span><span class="o">/</span><span class="n">vol1</span><span class="o">/</span><span class="n">fastq</span><span class="o">/</span><span class="n">SRR258</span><span class="o">/</span><span class="mi">003</span><span class="o">/</span><span class="n">SRR2584863</span><span class="o">/</span><span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">curl</span> <span class="o">-</span><span class="n">O</span> <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">sra</span><span class="o">.</span><span class="n">ebi</span><span class="o">.</span><span class="n">ac</span><span class="o">.</span><span class="n">uk</span><span class="o">/</span><span class="n">vol1</span><span class="o">/</span><span class="n">fastq</span><span class="o">/</span><span class="n">SRR258</span><span class="o">/</span><span class="mi">003</span><span class="o">/</span><span class="n">SRR2584863</span><span class="o">/</span><span class="n">SRR2584863_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">curl</span> <span class="o">-</span><span class="n">O</span> <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">sra</span><span class="o">.</span><span class="n">ebi</span><span class="o">.</span><span class="n">ac</span><span class="o">.</span><span class="n">uk</span><span class="o">/</span><span class="n">vol1</span><span class="o">/</span><span class="n">fastq</span><span class="o">/</span><span class="n">SRR258</span><span class="o">/</span><span class="mi">006</span><span class="o">/</span><span class="n">SRR2584866</span><span class="o">/</span><span class="n">SRR2584866_1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">curl</span> <span class="o">-</span><span class="n">O</span> <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">sra</span><span class="o">.</span><span class="n">ebi</span><span class="o">.</span><span class="n">ac</span><span class="o">.</span><span class="n">uk</span><span class="o">/</span><span class="n">vol1</span><span class="o">/</span><span class="n">fastq</span><span class="o">/</span><span class="n">SRR258</span><span class="o">/</span><span class="mi">006</span><span class="o">/</span><span class="n">SRR2584866</span><span class="o">/</span><span class="n">SRR2584866_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<blockquote>
<div>The data comes in a compressed format, which is why there is a <code class="docutils literal notranslate"><span class="pre">.gz</span></code> at the end of the file names. This makes it faster to transfer, and allows it to take up less space on our computer. Let’s unzip one of the files so that we can look at the fastq format.</div></blockquote>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ gunzip SRR2584863_1.fastq.gz
</pre></div>
</div>
</div>
<div class="section" id="quality-control">
<h1>16. Quality Control<a class="headerlink" href="#quality-control" title="Permalink to this headline">¶</a></h1>
<p>We will now assess the quality of the sequence reads contained in our fastq files.</p>
<p><img alt="_images/var_calling_workflow_qc.png" src="_images/var_calling_workflow_qc.png" /></p>
<div class="section" id="details-on-the-fastq-format">
<h2>16.1. Details on the FASTQ format<a class="headerlink" href="#details-on-the-fastq-format" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><p>Although it looks complicated (and it is), we can understand the <a class="reference external" href="https://en.wikipedia.org/wiki/FASTQ_format">fastq</a> format with a little decoding. Some rules about the format include…</p>
<ol class="simple">
<li>Always begins with ‘&#64;’ and then information about the read</li>
<li>The actual DNA sequence</li>
<li>Always begins with a ‘+’ and sometimes the same info in line 1</li>
<li>Has a string of characters which represent the quality scores; must have same number of characters as line 2</li>
</ol>
</div></blockquote>
<ul class="simple">
<li><a class="reference external" href="http://en.wikipedia.org/wiki/FASTQ_format">FASTQ Wikipedia</a></li>
</ul>
<p>We can view the first complete read in one of the files our dataset by using <code class="docutils literal notranslate"><span class="pre">head</span></code> to look at the first four lines.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ head -n 4 SRR2584863_1.fastq
</pre></div>
</div>
<blockquote>
<div>&#64;SRR2584863.1 HWI-ST957:244:H73TDADXX:1:1101:4712:2181/1
TTCACATCCTGACCATTCAGTTGAGCAAAATAGTTCTTCAGTGCCTGTTTAACCGAGTCACGCAGGGGTTTTTGGGTTACCTGATCCTGAGAGTTAACGGTAGAAACGGTCAGTACGTCAGAATTTACGCGTTGTTCGAACATAGTTCTG
+
CCCFFFFFGHHHHJIJJJJIJJJIIJJJJIIIJJGFIIIJEDDFEGGJIFHHJIJJDECCGGEGIIJFHFFFACD:BBBDDACCCCAA&#64;&#64;CA&#64;C&gt;C3&gt;&#64;5(8&amp;&gt;C:9?8+89&lt;4(:83825C(:A#########################</div></blockquote>
<blockquote>
<div><p><strong>Exercise</strong></p>
<p>What is the last read in the <code class="docutils literal notranslate"><span class="pre">SRR2584863_1.fastq</span></code> file? How confident are you in this read?</p>
</div></blockquote>
</div>
<div class="section" id="assessing-quality-using-fastqc">
<h2>16.2. Assessing Quality using FastQC<a class="headerlink" href="#assessing-quality-using-fastqc" title="Permalink to this headline">¶</a></h2>
<p>In real life, you won’t be assessing the quality of your reads by visually inspecting your FASTQ files. Rather, you’ll be using a software program to assess read quality and filter out poor quality reads. We’ll first use a program called <a class="reference external" href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> to visualize the quality of our reads.</p>
<p>FastQC has a number of features which can give you a quick impression of any problems your data may have, so you can take these issues into consideration before moving forward with your analyses. Rather than looking at quality scores for each individual read, FastQC looks at quality collectively across all reads within a sample. The image below shows one FastQC-generated plot that indicates a very high quality sample:</p>
<p><img alt="_images/fastqc_good.png" src="_images/fastqc_good.png" /></p>
<p>The x-axis displays the base position in the read, and the y-axis shows quality scores. In this example, the sample contains reads that are 40 bp long. This is much shorter than the reads we are working with in our workflow. For each position, there is a box-and-whisker plot showing
the distribution of quality scores for all reads at that position. The horizontal red line indicates the median quality score and the yellow box shows the 2nd to 3rd quartile range. This means that 50% of reads have a quality score that falls within the range of the yellow box at that position. The whiskers show the range to the 1st and 4th quartile.</p>
<p>For each position in this sample, the quality values do not drop much lower than 32. This is a high quality score. The plot background is also color-coded to identify good (green), acceptable (yellow), and bad (red) quality scores.</p>
<p>Now let’s take a look at a quality plot on the other end of the spectrum.</p>
<p><img alt="_images/fastqc_bad.png" src="_images/fastqc_bad.png" /></p>
<p>Here, we see positions within the read in which the boxes span a much wider range. Also, quality scores drop quite low into the “bad” range, particularly on the tail end of the reads. The FastQC tool produces several other diagnostic plots to assess sample quality, in addition to the one plotted above.</p>
</div>
<div class="section" id="running-fastqc">
<h2>16.3. Running FastQC<a class="headerlink" href="#running-fastqc" title="Permalink to this headline">¶</a></h2>
<p><strong>Note:</strong> If you haven’t added conda to your PATH yet, please do so by running the following:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>echo export PATH=$PATH:/opt/miniconda3/bin &gt;&gt; ~/.bashrc
</pre></div>
</div>
<p>Then, run the following command (or start a new terminal session) in order to activate the conda environment:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">source</span> <span class="o">~/.</span><span class="n">bashrc</span>
</pre></div>
</div>
<p>We will now assess the quality of the reads that we downloaded. First, make sure you’re still in the <code class="docutils literal notranslate"><span class="pre">untrimmed_fastq</span></code> directory with the <code class="docutils literal notranslate"><span class="pre">pwd</span></code> command if not, change directory using:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd ~/dc_workshop/data/untrimmed_fastq/
</pre></div>
</div>
<blockquote>
<div><p><strong>Exercise</strong></p>
<p>How big are the files?
(Hint: Look at the options for the <code class="docutils literal notranslate"><span class="pre">ls</span></code> command to see how to show file sizes.)</p>
</div></blockquote>
<p>FastQC can accept multiple file names as input, and on both zipped and unzipped files, so we can use the *.fastq* wildcard to run FastQC on all of the FASTQ files in this directory.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ fastqc *.fastq*
</pre></div>
</div>
<p>You will see an automatically updating output message telling you the progress of the analysis. It will start like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Started analysis of SRR2584863_1.fastq
Approx 5% complete for SRR2584863_1.fastq
Approx 10% complete for SRR2584863_1.fastq
Approx 15% complete for SRR2584863_1.fastq
Approx 20% complete for SRR2584863_1.fastq
Approx 25% complete for SRR2584863_1.fastq
Approx 30% complete for SRR2584863_1.fastq
Approx 35% complete for SRR2584863_1.fastq
Approx 40% complete for SRR2584863_1.fastq
Approx 45% complete for SRR2584863_1.fastq
....
Approx 80% complete for SRR2589044_2.fastq.gz
Approx 85% complete for SRR2589044_2.fastq.gz
Approx 90% complete for SRR2589044_2.fastq.gz
Approx 95% complete for SRR2589044_2.fastq.gz
Analysis complete for SRR2589044_2.fastq.gz
$
</pre></div>
</div>
<p>For each input FASTQ file, FastQC has created a <code class="docutils literal notranslate"><span class="pre">.zip</span></code> file and a <code class="docutils literal notranslate"><span class="pre">.html</span></code> file. The <code class="docutils literal notranslate"><span class="pre">.zip</span></code> file extension indicates that this is
actually a compressed set of multiple output files. We’ll be working with these output files soon. The <code class="docutils literal notranslate"><span class="pre">.html</span></code> file is a stable web-page
displaying the summary report for each of our samples.</p>
<p>We want to keep our data files and our results files separate, so we will move these output files into a new directory within our <code class="docutils literal notranslate"><span class="pre">results/</span></code> directory.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ mkdir -p ~/dc_workshop/results/fastqc_untrimmed_reads
$ mv *.zip ~/dc_workshop/results/fastqc_untrimmed_reads/
$ mv *.html ~/dc_workshop/results/fastqc_untrimmed_reads/
</pre></div>
</div>
<p>Now we can navigate into this results directory and do some closer inspection of our output files.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd ~/dc_workshop/results/fastqc_untrimmed_reads/
</pre></div>
</div>
<div class="section" id="multiqc">
<h3>16.3.1. MultiQC<a class="headerlink" href="#multiqc" title="Permalink to this headline">¶</a></h3>
<p>If you would like to aggregate all of your fastqc reports across many samples, <a class="reference external" href="http://multiqc.info/">MultiQC</a> will do this into a single report for easy comparison.</p>
<p>Run MultiQC:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">multiqc</span> <span class="o">.</span>
</pre></div>
</div>
<blockquote>
<div><strong>Note</strong> The <code class="docutils literal notranslate"><span class="pre">.</span></code> which follows the multiqc command is UNIX representation for the current folder.</div></blockquote>
<p>The terminal output should look like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">INFO</span>   <span class="p">]</span>         <span class="n">multiqc</span> <span class="p">:</span> <span class="n">This</span> <span class="ow">is</span> <span class="n">MultiQC</span> <span class="n">v1</span><span class="o">.</span><span class="mi">7</span>
<span class="p">[</span><span class="n">INFO</span>   <span class="p">]</span>         <span class="n">multiqc</span> <span class="p">:</span> <span class="n">Template</span>    <span class="p">:</span> <span class="n">default</span>
<span class="p">[</span><span class="n">INFO</span>   <span class="p">]</span>         <span class="n">multiqc</span> <span class="p">:</span> <span class="n">Searching</span> <span class="s1">&#39;.&#39;</span>
<span class="p">[</span><span class="n">INFO</span>   <span class="p">]</span>          <span class="n">fastqc</span> <span class="p">:</span> <span class="n">Found</span> <span class="mi">8</span> <span class="n">reports</span>
<span class="p">[</span><span class="n">INFO</span>   <span class="p">]</span>         <span class="n">multiqc</span> <span class="p">:</span> <span class="n">Compressing</span> <span class="n">plot</span> <span class="n">data</span>
<span class="p">[</span><span class="n">INFO</span>   <span class="p">]</span>         <span class="n">multiqc</span> <span class="p">:</span> <span class="n">Report</span>      <span class="p">:</span> <span class="n">multiqc_report</span><span class="o">.</span><span class="n">html</span>
<span class="p">[</span><span class="n">INFO</span>   <span class="p">]</span>         <span class="n">multiqc</span> <span class="p">:</span> <span class="n">Data</span>        <span class="p">:</span> <span class="n">multiqc_data</span>
<span class="p">[</span><span class="n">INFO</span>   <span class="p">]</span>         <span class="n">multiqc</span> <span class="p">:</span> <span class="n">MultiQC</span> <span class="n">complete</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="viewing-the-fastqc-results">
<h2>16.4. Viewing the FastQC results<a class="headerlink" href="#viewing-the-fastqc-results" title="Permalink to this headline">¶</a></h2>
<p>If we were working on our local computers, we’d be able to display each of these HTML files as a webpage:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ open SRR2584863_1_fastqc.html
</pre></div>
</div>
<p>However, if you try this on your remote instance, you’ll get an error:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Couldn</span><span class="s1">&#39;t get a file descriptor referring to the console</span>
</pre></div>
</div>
<p>This is because the remote instance we’re using doesn’t have any web browsers installed on it, so the remote computer doesn’t know how to
open the file. We want to look at the webpage summary reports, so let’s transfer them to our local computers (i.e. your laptop).</p>
<p>To transfer a file from a remote server to our own machines, we will use <code class="docutils literal notranslate"><span class="pre">scp</span></code>, which we learned yesterday in the Shell Genomics lesson.</p>
<p>First we will make a new directory on our computer to store the HTML files we’re transfering. Let’s put it on our desktop for now. Open a new
tab in your terminal program (you can use the pull down menu at the top of your screen or the <kbd>Cmd</kbd>+<kbd>t</kbd> keyboard shortcut) and type:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ mkdir -p ~/Desktop/fastqc_html
</pre></div>
</div>
<p>Now we can transfer our HTML files to our local computer using <code class="docutils literal notranslate"><span class="pre">scp</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ scp sateeshp@128.196.142.26:/home/sateeshp/dcgenomics/results/fastqc_untrimmed_reads/*.html ~/Desktop/fastqc_html/
</pre></div>
</div>
<ul class="simple">
<li>The first part of the command <code class="docutils literal notranslate"><span class="pre">sateeshp&#64;128.196.142.26</span></code> is the address for your remote computer.</li>
<li>The second part starts with a <code class="docutils literal notranslate"><span class="pre">:</span></code> and then gives the absolute path of the files you want to transfer from your remote computer. Don’t
forget the <code class="docutils literal notranslate"><span class="pre">:</span></code>. We used a wildcard (<code class="docutils literal notranslate"><span class="pre">*.html</span></code>) to indicate that we want all of the HTML files.</li>
<li>The third part of the command gives the absolute path of the location you want to put the files. This is on your local computer and is the
directory we just created <code class="docutils literal notranslate"><span class="pre">~/Desktop/fastqc_html</span></code>.</li>
</ul>
<p>You should see a status output like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">SRR097977_fastqc</span><span class="o">.</span><span class="n">html</span>                       <span class="mi">100</span><span class="o">%</span>  <span class="mi">576</span><span class="n">KB</span>   <span class="mf">1.6</span><span class="n">MB</span><span class="o">/</span><span class="n">s</span>   <span class="mi">00</span><span class="p">:</span><span class="mi">00</span>    
<span class="n">SRR098026_fastqc</span><span class="o">.</span><span class="n">html</span>                       <span class="mi">100</span><span class="o">%</span>  <span class="mi">620</span><span class="n">KB</span>   <span class="mf">3.3</span><span class="n">MB</span><span class="o">/</span><span class="n">s</span>   <span class="mi">00</span><span class="p">:</span><span class="mi">00</span>    
<span class="n">SRR2584863_1_fastqc</span><span class="o">.</span><span class="n">html</span>                    <span class="mi">100</span><span class="o">%</span>  <span class="mi">626</span><span class="n">KB</span>   <span class="mf">3.1</span><span class="n">MB</span><span class="o">/</span><span class="n">s</span>   <span class="mi">00</span><span class="p">:</span><span class="mi">00</span>    
<span class="n">SRR2584863_2_fastqc</span><span class="o">.</span><span class="n">html</span>                    <span class="mi">100</span><span class="o">%</span>  <span class="mi">632</span><span class="n">KB</span>   <span class="mf">3.1</span><span class="n">MB</span><span class="o">/</span><span class="n">s</span>   <span class="mi">00</span><span class="p">:</span><span class="mi">00</span>    
<span class="n">SRR2584866_1_fastqc</span><span class="o">.</span><span class="n">html</span>                    <span class="mi">100</span><span class="o">%</span>  <span class="mi">631</span><span class="n">KB</span>   <span class="mf">3.2</span><span class="n">MB</span><span class="o">/</span><span class="n">s</span>   <span class="mi">00</span><span class="p">:</span><span class="mi">00</span>    
<span class="n">SRR2584866_2_fastqc</span><span class="o">.</span><span class="n">html</span>                    <span class="mi">100</span><span class="o">%</span>  <span class="mi">626</span><span class="n">KB</span>   <span class="mf">3.6</span><span class="n">MB</span><span class="o">/</span><span class="n">s</span>   <span class="mi">00</span><span class="p">:</span><span class="mi">00</span>    
<span class="n">SRR2589044_1_fastqc</span><span class="o">.</span><span class="n">html</span>                    <span class="mi">100</span><span class="o">%</span>  <span class="mi">626</span><span class="n">KB</span>   <span class="mf">3.7</span><span class="n">MB</span><span class="o">/</span><span class="n">s</span>   <span class="mi">00</span><span class="p">:</span><span class="mi">00</span>    
<span class="n">SRR2589044_2_fastqc</span><span class="o">.</span><span class="n">html</span>                    <span class="mi">100</span><span class="o">%</span>  <span class="mi">627</span><span class="n">KB</span>   <span class="mf">3.1</span><span class="n">MB</span><span class="o">/</span><span class="n">s</span>   <span class="mi">00</span><span class="p">:</span><span class="mi">00</span>    
<span class="n">multiqc_report</span><span class="o">.</span><span class="n">html</span>                         <span class="mi">100</span><span class="o">%</span> <span class="mi">1174</span><span class="n">KB</span>   <span class="mf">4.0</span><span class="n">MB</span><span class="o">/</span><span class="n">s</span>   <span class="mi">00</span><span class="p">:</span><span class="mi">00</span>
</pre></div>
</div>
<blockquote>
<div><p><strong>Exercise</strong></p>
<p>Discuss your results with a neighbor. Which sample(s) looks the best in terms of per base sequence quality? Which sample(s) look the worst?</p>
<blockquote>
<div><em>Solution</em>
All of the reads contain usable data, but the quality decreases toward the end of the reads.</div></blockquote>
</div></blockquote>
</div>
<div class="section" id="decoding-the-other-fastqc-outputs">
<h2>16.5. Decoding the other FastQC outputs<a class="headerlink" href="#decoding-the-other-fastqc-outputs" title="Permalink to this headline">¶</a></h2>
<p>We’ve now looked at quite a few “Per base sequence quality” FastQC graphs, but there are nine other graphs that we haven’t talked about! Below we have provided a brief overview of interpretations for each of these plots. It’s important to keep in mind</p>
<ul class="simple">
<li><strong>Per tile sequence quality</strong>: the machines that perform sequencing are divided into tiles. This plot displays patterns in base quality along these tiles. Consistently low scores are often found around the edges, but hot spots can also occur in the middle if an air bubble was introduced at some point during the run.</li>
<li><strong>Per sequence quality scores</strong>: a density plot of quality for all reads at all positions. This plot shows what quality scores are most common.</li>
<li><strong>Per base sequence content</strong>: plots the proportion of each base position over all of the reads. Typically, we expect to see each base roughly 25% of the time at each position, but this often fails at the beginning or end of the read due to quality or adapter content.</li>
<li><strong>Per sequence GC content</strong>: a density plot of average GC content in each of the reads.</li>
<li><strong>Per base N content</strong>: the percent of times that ‘N’ occurs at a position in all reads. If there is an increase at a particular position, this might indicate that something went wrong during sequencing.</li>
<li><strong>Sequence Length Distribution</strong>: the distribution of sequence lengths of all reads in the file. If the data is raw, there is often on sharp peak, however if the reads have been trimmed, there may be a distribution of shorter lengths.</li>
<li><strong>Sequence Duplication Levels</strong>: A distribution of duplicated sequences. In sequencing, we expect most reads to only occur once. If some sequences are occurring more than once, it might indicate enrichment bias (e.g. from PCR). If the samples are high coverage (or RNA-seq or amplicon), this might not be true.</li>
<li><strong>Overrepresented sequences</strong>: A list of sequences that occur more frequently than would be expected by chance.</li>
<li><strong>Adapter Content</strong>: a graph indicating where adapter sequences occur in the reads.</li>
</ul>
</div>
</div>


          </div>
            
  <div class="footer-relations">
    
      <div class="pull-left">
        <a class="btn btn-default" href="workflow_overview.html" title="previous chapter (use the left arrow)">12. Workflow_Overview</a>
      </div>
    
      <div class="pull-right">
        <a class="btn btn-default" href="trimming.html" title="next chapter (use the right arrow)">17. Trimming and Filtering</a>
      </div>
    </div>
    <div class="clearer"></div>
  
        </div>
        <div class="clearfix"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="trimming.html" title="17. Trimming and Filtering"
             >next</a> |</li>
        <li class="right" >
          <a href="workflow_overview.html" title="12. Workflow_Overview"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="toc.html">2019-04-11-gwu 0.0.1 documentation</a> &#187;</li> 
      </ul>
    </div>
<script type="text/javascript">
  $("#mobile-toggle a").click(function () {
    $("#left-column").toggle();
  });
</script>
<script type="text/javascript" src="_static/js/bootstrap.js"></script>
  <div class="footer">
    &copy; Copyright 2019, Sateesh Peri. Created using <a href="http://sphinx.pocoo.org/">Sphinx</a>.
  </div>
  </body>
</html>