
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  
  <!-- Licensed under the Apache 2.0 License -->
  <link rel="stylesheet" type="text/css" href="_static/fonts/open-sans/stylesheet.css" />
  <!-- Licensed under the SIL Open Font License -->
  <link rel="stylesheet" type="text/css" href="_static/fonts/source-serif-pro/source-serif-pro.css" />
  <link rel="stylesheet" type="text/css" href="_static/css/bootstrap.min.css" />
  <link rel="stylesheet" type="text/css" href="_static/css/bootstrap-theme.min.css" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
    <title>12. Assessing Read Quality &#8212; 2019-04-11-gwu 0.0.1 documentation</title>
    <link rel="stylesheet" href="_static/guzzle.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="17. Trimming and Filtering" href="trimming.html" />
    <link rel="prev" title="11. Working with Bioconda." href="bioconda_lesson.html" />
  
   

  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="trimming.html" title="17. Trimming and Filtering"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="bioconda_lesson.html" title="11. Working with Bioconda."
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="toc.html">2019-04-11-gwu 0.0.1 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="container-wrapper">

      <div id="mobile-toggle">
        <a href="#"><span class="glyphicon glyphicon-align-justify" aria-hidden="true"></span></a>
      </div>
  <div id="left-column">
    <div class="sphinxsidebar">
<div class="sidebar-block">
  <div class="sidebar-wrapper">
    <h2>Contents</h2>
    <div class="sidebar-localtoc">
      <ul>
<li><a class="reference internal" href="#">12. Assessing Read Quality</a></li>
<li><a class="reference internal" href="#bioinformatics-workflows">13. Bioinformatics workflows</a></li>
<li><a class="reference internal" href="#starting-with-data">14. Starting with Data</a></li>
<li><a class="reference internal" href="#quality-control">15. Quality Control</a><ul>
<li><a class="reference internal" href="#details-on-the-fastq-format">15.1. Details on the FASTQ format</a></li>
<li><a class="reference internal" href="#assessing-quality-using-fastqc">15.2. Assessing Quality using FastQC</a></li>
<li><a class="reference internal" href="#running-fastqc">15.3. Running FastQC</a></li>
<li><a class="reference internal" href="#viewing-the-fastqc-results">15.4. Viewing the FastQC results</a></li>
<li><a class="reference internal" href="#decoding-the-other-fastqc-outputs">15.5. Decoding the other FastQC outputs</a></li>
<li><a class="reference internal" href="#working-with-the-fastqc-text-output">15.6. Working with the FastQC text output</a></li>
<li><a class="reference internal" href="#documenting-our-work">15.7. Documenting Our Work</a></li>
</ul>
</li>
<li><a class="reference internal" href="#other-notes-optional">16. Other notes  – Optional</a></li>
</ul>

    </div>
  </div>
</div>
  <h4>Previous topic</h4>
  <p class="topless"><a href="bioconda_lesson.html"
                        title="previous chapter">11. Working with Bioconda.</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="trimming.html"
                        title="next chapter">17. Trimming and Filtering</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/quality_control.md.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div class="sidebar-block">
  <div class="sidebar-wrapper">
    <div id="main-search">
      <form class="form-inline" action="search.html" method="GET" role="form">
        <div class="input-group">
          <input name="q" type="text" class="form-control" placeholder="Search...">
        </div>
        <input type="hidden" name="check_keywords" value="yes" />
        <input type="hidden" name="area" value="default" />
      </form>
    </div>
  </div>
</div>
      
    </div>
  </div>
        <div id="right-column">
          
          <div role="navigation" aria-label="breadcrumbs navigation">
            <ol class="breadcrumb">
              <li><a href="toc.html">Docs</a></li>
              
              <li>12. Assessing Read Quality</li>
            </ol>
          </div>
          
          <div class="document clearer body">
            
  <div class="section" id="assessing-read-quality">
<h1>12. Assessing Read Quality<a class="headerlink" href="#assessing-read-quality" title="Permalink to this headline">¶</a></h1>
<blockquote>
<div><p>objectives:</p>
<blockquote>
<div><ul class="simple">
<li>“Explain how a FASTQ file encodes per-base quality scores.”</li>
<li>“Interpret a FastQC plot summarizing per-base quality across all reads.”</li>
<li>“Use <code class="docutils literal notranslate"><span class="pre">for</span></code> loops to automate operations on multiple files.”</li>
</ul>
</div></blockquote>
</div></blockquote>
<blockquote>
<div><p>keypoints:</p>
<blockquote>
<div><ul class="simple">
<li>“Quality encodings vary across sequencing platforms.”</li>
<li>“<code class="docutils literal notranslate"><span class="pre">for</span></code> loops let you perform the same set of operations on multiple files with a single command.”</li>
</ul>
</div></blockquote>
</div></blockquote>
</div>
<hr class="docutils" />
<div class="section" id="bioinformatics-workflows">
<h1>13. Bioinformatics workflows<a class="headerlink" href="#bioinformatics-workflows" title="Permalink to this headline">¶</a></h1>
<p>When working with high-throughput sequencing data, the raw reads you get off of the sequencer will need to pass
through a number of  different tools in order to generate your final desired output. The execution of this set of
tools in a specified order is commonly referred to as a <em>workflow</em> or a <em>pipeline</em>.</p>
<p>An example of the workflow we will be using for our variant calling analysis is provided below with a brief
description of each step.</p>
<p><img alt="_images/variant_calling_workflow.png" src="_images/variant_calling_workflow.png" /></p>
<ol class="simple">
<li>Quality control - Assessing quality using FastQC</li>
<li>Quality control - Trimming and/or filtering reads (if necessary)</li>
<li>Align reads to reference genome</li>
<li>Perform post-alignment clean-up</li>
<li>Variant calling</li>
</ol>
<p>These workflows in bioinformatics adopt a plug-and-play approach in that the output of one tool can be easily
used as input to another tool without any extensive configuration. Having standards for data formats is what
makes this feasible. Standards ensure that data is stored in a way that is generally accepted and agreed upon
within the community. The tools that are used to analyze data at different stages of the workflow are therefore
built under the assumption that the data will be provided in a specific format.</p>
</div>
<div class="section" id="starting-with-data">
<h1>14. Starting with Data<a class="headerlink" href="#starting-with-data" title="Permalink to this headline">¶</a></h1>
<p>Often times, the first step in a bioinformatics workflow is getting the data you want to work with onto a computer where you can work with it. If you have sequenced your own data, the sequencing center will usually provide you with a link that you can use to download your data. Today we will be working with publicly available sequencing data.</p>
<p>We are studying a population of <em>Escherichia coli</em> (designated Ara-3), which were propagated for more than 50,000 generations in a glucose-limited minimal medium. We will be working with three samples from this experiment, one from 5,000 generations, one from 15,000 generations, and one from 50,000 generations. The population changed substantially during the course of the experiment, and we will be exploring how with our variant calling workflow.</p>
<p>The data are paired-end, so we will download two files for each sample. We will use the <a class="reference external" href="https://www.ebi.ac.uk/ena">European Nucleotide Archive</a> to get our data. The ENA “provides a comprehensive record of the world’s nucleotide sequencing information, covering raw sequencing data, sequence assembly information and functional annotation.” The ENA also provides sequencing data in the fastq format, an important format for sequencing reads that we will be learning about today.</p>
<p>To download the data, run the commands below. It will take about 10 minutes to download the files.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mkdir</span> <span class="o">-</span><span class="n">p</span> <span class="o">~/</span><span class="n">dc_workshop</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">untrimmed_fastq</span><span class="o">/</span>
<span class="n">cd</span> <span class="o">~/</span><span class="n">dc_workshop</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">untrimmed_fastq</span>

<span class="n">curl</span> <span class="o">-</span><span class="n">O</span> <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">sra</span><span class="o">.</span><span class="n">ebi</span><span class="o">.</span><span class="n">ac</span><span class="o">.</span><span class="n">uk</span><span class="o">/</span><span class="n">vol1</span><span class="o">/</span><span class="n">fastq</span><span class="o">/</span><span class="n">SRR258</span><span class="o">/</span><span class="mi">004</span><span class="o">/</span><span class="n">SRR2589044</span><span class="o">/</span><span class="n">SRR2589044_1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">curl</span> <span class="o">-</span><span class="n">O</span> <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">sra</span><span class="o">.</span><span class="n">ebi</span><span class="o">.</span><span class="n">ac</span><span class="o">.</span><span class="n">uk</span><span class="o">/</span><span class="n">vol1</span><span class="o">/</span><span class="n">fastq</span><span class="o">/</span><span class="n">SRR258</span><span class="o">/</span><span class="mi">004</span><span class="o">/</span><span class="n">SRR2589044</span><span class="o">/</span><span class="n">SRR2589044_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">curl</span> <span class="o">-</span><span class="n">O</span> <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">sra</span><span class="o">.</span><span class="n">ebi</span><span class="o">.</span><span class="n">ac</span><span class="o">.</span><span class="n">uk</span><span class="o">/</span><span class="n">vol1</span><span class="o">/</span><span class="n">fastq</span><span class="o">/</span><span class="n">SRR258</span><span class="o">/</span><span class="mi">003</span><span class="o">/</span><span class="n">SRR2584863</span><span class="o">/</span><span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">curl</span> <span class="o">-</span><span class="n">O</span> <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">sra</span><span class="o">.</span><span class="n">ebi</span><span class="o">.</span><span class="n">ac</span><span class="o">.</span><span class="n">uk</span><span class="o">/</span><span class="n">vol1</span><span class="o">/</span><span class="n">fastq</span><span class="o">/</span><span class="n">SRR258</span><span class="o">/</span><span class="mi">003</span><span class="o">/</span><span class="n">SRR2584863</span><span class="o">/</span><span class="n">SRR2584863_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">curl</span> <span class="o">-</span><span class="n">O</span> <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">sra</span><span class="o">.</span><span class="n">ebi</span><span class="o">.</span><span class="n">ac</span><span class="o">.</span><span class="n">uk</span><span class="o">/</span><span class="n">vol1</span><span class="o">/</span><span class="n">fastq</span><span class="o">/</span><span class="n">SRR258</span><span class="o">/</span><span class="mi">006</span><span class="o">/</span><span class="n">SRR2584866</span><span class="o">/</span><span class="n">SRR2584866_1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">curl</span> <span class="o">-</span><span class="n">O</span> <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">sra</span><span class="o">.</span><span class="n">ebi</span><span class="o">.</span><span class="n">ac</span><span class="o">.</span><span class="n">uk</span><span class="o">/</span><span class="n">vol1</span><span class="o">/</span><span class="n">fastq</span><span class="o">/</span><span class="n">SRR258</span><span class="o">/</span><span class="mi">006</span><span class="o">/</span><span class="n">SRR2584866</span><span class="o">/</span><span class="n">SRR2584866_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>The data comes in a compressed format, which is why there is a <code class="docutils literal notranslate"><span class="pre">.gz</span></code> at the end of the file names. This makes it faster to transfer, and allows it to take up less space on our computer. Let’s unzip one of the files so that we can look at the fastq format.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ gunzip SRR2584863_1.fastq.gz
</pre></div>
</div>
</div>
<div class="section" id="quality-control">
<h1>15. Quality Control<a class="headerlink" href="#quality-control" title="Permalink to this headline">¶</a></h1>
<p>We will now assess the quality of the sequence reads contained in our fastq files.</p>
<p><img alt="_images/var_calling_workflow_qc.png" src="_images/var_calling_workflow_qc.png" /></p>
<div class="section" id="details-on-the-fastq-format">
<h2>15.1. Details on the FASTQ format<a class="headerlink" href="#details-on-the-fastq-format" title="Permalink to this headline">¶</a></h2>
<p>Although it looks complicated (and it is), we can understand the
<a class="reference external" href="https://en.wikipedia.org/wiki/FASTQ_format">fastq</a> format with a little decoding. Some rules about the format
include…</p>
<ol class="simple">
<li>Always begins with ‘&#64;’ and then information about the read</li>
<li>The actual DNA sequence</li>
<li>Always begins with a ‘+’ and sometimes the same info in line 1</li>
<li>Has a string of characters which represent the quality scores; must have same number of characters as line 2</li>
</ol>
<p>We can view the first complete read in one of the files our dataset by using <code class="docutils literal notranslate"><span class="pre">head</span></code> to look at the first four lines.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ head -n 4 SRR2584863_1.fastq
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>@SRR2584863.1 HWI-ST957:244:H73TDADXX:1:1101:4712:2181/1
TTCACATCCTGACCATTCAGTTGAGCAAAATAGTTCTTCAGTGCCTGTTTAACCGAGTCACGCAGGGGTTTTTGGGTTACCTGATCCTGAGAGTTAACGGTAGAAACGGTCAGTACGTCAGAATTTACGCGTTGTTCGAACATAGTTCTG
+
CCCFFFFFGHHHHJIJJJJIJJJIIJJJJIIIJJGFIIIJEDDFEGGJIFHHJIJJDECCGGEGIIJFHFFFACD:BBBDDACCCCAA@@CA@C&gt;C3&gt;@5(8&amp;&gt;C:9?8+89&lt;4(:83825C(:A#########################
</pre></div>
</div>
<p>Line 4 shows the quality for each nucleotide in the read. Quality is interpreted as the
probability of an incorrect base call (e.g. 1 in 10) or, equivalently, the base call
accuracy (e.g. 90%). To make it possible to line up each individual nucleotide with its quality
score, the numerical score is converted into a code where each individual character
represents the numerical quality score for an individual nucleotide. For example, in the line
above, the quality score line is:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>CCCFFFFFGHHHHJIJJJJIJJJIIJJJJIIIJJGFIIIJEDDFEGGJIFHHJIJJDECCGGEGIIJFHFFFACD:BBBDDACCCCAA@@CA@C&gt;C3&gt;@5(8&amp;&gt;C:9?8+89&lt;4(:83825C(:A#########################
</pre></div>
</div>
<p>The numerical value assigned to each of these characters depends on the
sequencing platform that generated the reads. The sequencing machine used to generate our data
uses the standard Sanger quality PHRED score encoding, using by Illumina version 1.8 onwards.
Each character is assigned a quality score between 0 and 40 as shown in the chart below.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Quality encoding: !&quot;#$%&amp;&#39;()*+,-./0123456789:;&lt;=&gt;?@ABCDEFGHI
                  |         |         |         |         |
Quality score:    0........10........20........30........40
</pre></div>
</div>
<p>Each quality score represents the probability that the corresponding nucleotide call is
incorrect. This quality score is logarithmically based, so a quality score of 10 reflects a
base call accuracy of 90%, but a quality score of 20 reflects a base call accuracy of 99%.
These probability values are the results from the base calling algorithm and dependent on how
much signal was captured for the base incorporation.</p>
<p>Looking back at our read:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>@SRR2584863.1 HWI-ST957:244:H73TDADXX:1:1101:4712:2181/1
TTCACATCCTGACCATTCAGTTGAGCAAAATAGTTCTTCAGTGCCTGTTTAACCGAGTCACGCAGGGGTTTTTGGGTTACCTGATCCTGAGAGTTAACGGTAGAAACGGTCAGTACGTCAGAATTTACGCGTTGTTCGAACATAGTTCTG
+
CCCFFFFFGHHHHJIJJJJIJJJIIJJJJIIIJJGFIIIJEDDFEGGJIFHHJIJJDECCGGEGIIJFHFFFACD:BBBDDACCCCAA@@CA@C&gt;C3&gt;@5(8&amp;&gt;C:9?8+89&lt;4(:83825C(:A#########################
</pre></div>
</div>
<p>we can now see that there are a range of quality score, but that the end of the sequence
very poor (<code class="docutils literal notranslate"><span class="pre">#</span></code> = a quality score of 2).</p>
<blockquote>
<div><p><strong>Exercise</strong></p>
<p>What is the last read in the <code class="docutils literal notranslate"><span class="pre">SRR2584863_1.fastq</span></code> file? How confident
are you in this read?</p>
<blockquote>
<div><p><em>Solution</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ tail -n 4 SRR2584863_1.fastq
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>@SRR2584863.1553259 HWI-ST957:245:H73R4ADXX:2:2216:21048:100894/1
CTGCAATACCACGCTGATCTTTCACATGATGTAAGAAAAGTGGGATCAGCAAACCGGGTGCTGCTGTGGCTAGTTGCAGCAAACCATGCAGTGAACCCGCCTGTGCTTCGCTATAGCCGTGACTGATGAGGATCGCCGGAAGCCAGCCAA
+
CCCFFFFFHHHHGJJJJJJJJJHGIJJJIJJJJIJJJJIIIIJJJJJJJJJJJJJIIJJJHHHHHFFFFFEEEEEDDDDDDDDDDDDDDDDDCDEDDBDBDDBDDDDDDDDDBDEEDDDD7@BDDDDDD&gt;AA&gt;?B?&lt;@BDD@BDC?BDA?
</pre></div>
</div>
<p>This read has more consistent quality at its end than the first
read that we looked at, but still has a range of quality scores,
most of them high. We will look at variations in position-based quality
in just a moment.</p>
</div></blockquote>
</div></blockquote>
</div>
<div class="section" id="assessing-quality-using-fastqc">
<h2>15.2. Assessing Quality using FastQC<a class="headerlink" href="#assessing-quality-using-fastqc" title="Permalink to this headline">¶</a></h2>
<p>In real life, you won’t be assessing the quality of your reads by visually inspecting your
FASTQ files. Rather, you’ll be using a software program to assess read quality and
filter out poor quality reads. We’ll first use a program called <a class="reference external" href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> to visualize the quality of our reads.
Later in our workflow, we’ll use another program to filter out poor quality reads.</p>
<p>FastQC has a number of features which can give you a quick impression of any problems your
data may have, so you can take these issues into consideration before moving forward with your
analyses. Rather than looking at quality scores for each individual read, FastQC looks at
quality collectively across all reads within a sample. The image below shows one FastQC-generated plot that indicates
a very high quality sample:</p>
<p><img alt="img/good_quality1.8.png" src="img/good_quality1.8.png" /></p>
<p>The x-axis displays the base position in the read, and the y-axis shows quality scores. In this
example, the sample contains reads that are 40 bp long. This is much shorter than the reads we
are working with in our workflow. For each position, there is a box-and-whisker plot showing
the distribution of quality scores for all reads at that position. The horizontal red line
indicates the median quality score and the yellow box shows the 2nd to
3rd quartile range. This means that 50% of reads have a quality score that falls within the
range of the yellow box at that position. The whiskers show the range to the 1st and 4th
quartile.</p>
<p>For each position in this sample, the quality values do not drop much lower than 32. This
is a high quality score. The plot background is also color-coded to identify good (green),
acceptable (yellow), and bad (red) quality scores.</p>
<p>Now let’s take a look at a quality plot on the other end of the spectrum.</p>
<p><img alt="img/bad_quality1.8.png" src="img/bad_quality1.8.png" /></p>
<p>Here, we see positions within the read in which the boxes span a much wider range. Also, quality scores drop quite low into the “bad” range, particularly on the tail end of the reads. The FastQC tool produces several other diagnostic plots to assess sample quality, in addition to the one plotted above.</p>
</div>
<div class="section" id="running-fastqc">
<h2>15.3. Running FastQC<a class="headerlink" href="#running-fastqc" title="Permalink to this headline">¶</a></h2>
<p>We will now assess the quality of the reads that we downloaded. First, make sure you’re still in the <code class="docutils literal notranslate"><span class="pre">untrimmed_fastq</span></code> directory</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd ~/dc_workshop/data/untrimmed_fastq/
</pre></div>
</div>
<blockquote>
<div><p><strong>Exercise</strong></p>
<p>How big are the files?
(Hint: Look at the options for the <code class="docutils literal notranslate"><span class="pre">ls</span></code> command to see how to show
file sizes.)</p>
<blockquote>
<div><p><em>Solution</em></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ls -l -h
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">dcuser</span> <span class="n">dcuser</span> <span class="mi">545</span><span class="n">M</span> <span class="n">Jul</span>  <span class="mi">6</span> <span class="mi">20</span><span class="p">:</span><span class="mi">27</span> <span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">dcuser</span> <span class="n">dcuser</span> <span class="mi">183</span><span class="n">M</span> <span class="n">Jul</span>  <span class="mi">6</span> <span class="mi">20</span><span class="p">:</span><span class="mi">29</span> <span class="n">SRR2584863_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">dcuser</span> <span class="n">dcuser</span> <span class="mi">309</span><span class="n">M</span> <span class="n">Jul</span>  <span class="mi">6</span> <span class="mi">20</span><span class="p">:</span><span class="mi">34</span> <span class="n">SRR2584866_1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">dcuser</span> <span class="n">dcuser</span> <span class="mi">296</span><span class="n">M</span> <span class="n">Jul</span>  <span class="mi">6</span> <span class="mi">20</span><span class="p">:</span><span class="mi">37</span> <span class="n">SRR2584866_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">dcuser</span> <span class="n">dcuser</span> <span class="mi">124</span><span class="n">M</span> <span class="n">Jul</span>  <span class="mi">6</span> <span class="mi">20</span><span class="p">:</span><span class="mi">22</span> <span class="n">SRR2589044_1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">dcuser</span> <span class="n">dcuser</span> <span class="mi">128</span><span class="n">M</span> <span class="n">Jul</span>  <span class="mi">6</span> <span class="mi">20</span><span class="p">:</span><span class="mi">24</span> <span class="n">SRR2589044_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>There are six FASTQ files ranging from 124M (124MB) to 545M.</p>
</div></blockquote>
</div></blockquote>
<p>FastQC can accept multiple file names as input, and on both zipped and unzipped files, so we can use the *.fastq* wildcard to run FastQC on all of the FASTQ files in this directory.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ fastqc *.fastq*
</pre></div>
</div>
<p>You will see an automatically updating output message telling you the
progress of the analysis. It will start like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Started</span> <span class="n">analysis</span> <span class="n">of</span> <span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">5</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">10</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">15</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">20</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">25</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">30</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">35</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">40</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">45</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span>
</pre></div>
</div>
<p>In total, it should take about five minutes for FastQC to run on all
six of our FASTQ files. When the analysis completes, your prompt
will return. So your screen will look something like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Approx 80% complete for SRR2589044_2.fastq.gz
Approx 85% complete for SRR2589044_2.fastq.gz
Approx 90% complete for SRR2589044_2.fastq.gz
Approx 95% complete for SRR2589044_2.fastq.gz
Analysis complete for SRR2589044_2.fastq.gz
$
</pre></div>
</div>
<p>The FastQC program has created several new files within our
<code class="docutils literal notranslate"><span class="pre">data/untrimmed_fastq/</span></code> directory.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ls
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span>        <span class="n">SRR2584866_1_fastqc</span><span class="o">.</span><span class="n">html</span>  <span class="n">SRR2589044_1_fastqc</span><span class="o">.</span><span class="n">html</span>
<span class="n">SRR2584863_1_fastqc</span><span class="o">.</span><span class="n">html</span>  <span class="n">SRR2584866_1_fastqc</span><span class="o">.</span><span class="n">zip</span>   <span class="n">SRR2589044_1_fastqc</span><span class="o">.</span><span class="n">zip</span>
<span class="n">SRR2584863_1_fastqc</span><span class="o">.</span><span class="n">zip</span>   <span class="n">SRR2584866_1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>     <span class="n">SRR2589044_1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">SRR2584863_2_fastqc</span><span class="o">.</span><span class="n">html</span>  <span class="n">SRR2584866_2_fastqc</span><span class="o">.</span><span class="n">html</span>  <span class="n">SRR2589044_2_fastqc</span><span class="o">.</span><span class="n">html</span>
<span class="n">SRR2584863_2_fastqc</span><span class="o">.</span><span class="n">zip</span>   <span class="n">SRR2584866_2_fastqc</span><span class="o">.</span><span class="n">zip</span>   <span class="n">SRR2589044_2_fastqc</span><span class="o">.</span><span class="n">zip</span>
<span class="n">SRR2584863_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>     <span class="n">SRR2584866_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>     <span class="n">SRR2589044_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>For each input FASTQ file, FastQC has created a <code class="docutils literal notranslate"><span class="pre">.zip</span></code> file and a
<code class="docutils literal notranslate"><span class="pre">.html</span></code> file. The <code class="docutils literal notranslate"><span class="pre">.zip</span></code> file extension indicates that this is
actually a compressed set of multiple output files. We’ll be working
with these output files soon. The <code class="docutils literal notranslate"><span class="pre">.html</span></code> file is a stable webpage
displaying the summary report for each of our samples.</p>
<p>We want to keep our data files and our results files separate, so we
will move these
output files into a new directory within our <code class="docutils literal notranslate"><span class="pre">results/</span></code> directory.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ mkdir -p ~/dc_workshop/results/fastqc_untrimmed_reads
$ mv *.zip ~/dc_workshop/results/fastqc_untrimmed_reads/
$ mv *.html ~/dc_workshop/results/fastqc_untrimmed_reads/
</pre></div>
</div>
<p>Now we can navigate into this results directory and do some closer
inspection of our output files.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd ~/dc_workshop/results/fastqc_untrimmed_reads/
</pre></div>
</div>
</div>
<div class="section" id="viewing-the-fastqc-results">
<h2>15.4. Viewing the FastQC results<a class="headerlink" href="#viewing-the-fastqc-results" title="Permalink to this headline">¶</a></h2>
<p>If we were working on our local computers, we’d be able to display each of these
HTML files as a webpage:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ open SRR2584863_1_fastqc.html
</pre></div>
</div>
<p>However, if you try this on our AWS instance, you’ll get an error:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Couldn</span><span class="s1">&#39;t get a file descriptor referring to the console</span>
</pre></div>
</div>
<p>This is because the AWS instance we’re using doesn’t have any web
browsers installed on it, so the remote computer doesn’t know how to
open the file. We want to look at the webpage summary reports, so
let’s transfer them to our local computers (i.e. your laptop).</p>
<p>To transfer a file from a remote server to our own machines, we will
use <code class="docutils literal notranslate"><span class="pre">scp</span></code>, which we learned yesterday in the Shell Genomics lesson.</p>
<p>First we
will make a new directory on our computer to store the HTML files
we’re transfering. Let’s put it on our desktop for now. Open a new
tab in your terminal program (you can use the pull down menu at the
top of your screen or the Cmd+t keyboard shortcut) and type:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ mkdir -p ~/Desktop/fastqc_html
</pre></div>
</div>
<p>Now we can transfer our HTML files to our local computer using <code class="docutils literal notranslate"><span class="pre">scp</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ scp dcuser@ec2-34-238-162-94.compute-1.amazonaws.com:~/dc_workshop/results/fastqc_untrimmed_reads/*.html ~/Desktop/fastqc_html
</pre></div>
</div>
<p>As a reminder, the first part
of the command <code class="docutils literal notranslate"><span class="pre">dcuser&#64;ec2-34-238-162-94.compute-1.amazonaws.com</span></code> is
the address for your remote computer. Make sure you replace everything
after <code class="docutils literal notranslate"><span class="pre">dcuser&#64;</span></code> with your instance number (the one you used to log in).</p>
<p>The second part starts with a <code class="docutils literal notranslate"><span class="pre">:</span></code> and then gives the absolute path
of the files you want to transfer from your remote computer. Don’t
forget the <code class="docutils literal notranslate"><span class="pre">:</span></code>. We used a wildcard (<code class="docutils literal notranslate"><span class="pre">*.html</span></code>) to indicate that we want all of
the HTML files.</p>
<p>The third part of the command gives the absolute path of the location
you want to put the files. This is on your local computer and is the
directory we just created <code class="docutils literal notranslate"><span class="pre">~/Desktop/fastqc_html</span></code>.</p>
<p>You should see a status output like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">SRR2584863_1_fastqc</span><span class="o">.</span><span class="n">html</span>                      <span class="mi">100</span><span class="o">%</span>  <span class="mi">249</span><span class="n">KB</span> <span class="mf">152.3</span><span class="n">KB</span><span class="o">/</span><span class="n">s</span>   <span class="mi">00</span><span class="p">:</span><span class="mi">01</span>
<span class="n">SRR2584863_2_fastqc</span><span class="o">.</span><span class="n">html</span>                      <span class="mi">100</span><span class="o">%</span>  <span class="mi">254</span><span class="n">KB</span> <span class="mf">219.8</span><span class="n">KB</span><span class="o">/</span><span class="n">s</span>   <span class="mi">00</span><span class="p">:</span><span class="mi">01</span>
<span class="n">SRR2584866_1_fastqc</span><span class="o">.</span><span class="n">html</span>                      <span class="mi">100</span><span class="o">%</span>  <span class="mi">254</span><span class="n">KB</span> <span class="mf">271.8</span><span class="n">KB</span><span class="o">/</span><span class="n">s</span>   <span class="mi">00</span><span class="p">:</span><span class="mi">00</span>
<span class="n">SRR2584866_2_fastqc</span><span class="o">.</span><span class="n">html</span>                      <span class="mi">100</span><span class="o">%</span>  <span class="mi">251</span><span class="n">KB</span> <span class="mf">252.8</span><span class="n">KB</span><span class="o">/</span><span class="n">s</span>   <span class="mi">00</span><span class="p">:</span><span class="mi">00</span>
<span class="n">SRR2589044_1_fastqc</span><span class="o">.</span><span class="n">html</span>                      <span class="mi">100</span><span class="o">%</span>  <span class="mi">249</span><span class="n">KB</span> <span class="mf">370.1</span><span class="n">KB</span><span class="o">/</span><span class="n">s</span>   <span class="mi">00</span><span class="p">:</span><span class="mi">00</span>
<span class="n">SRR2589044_2_fastqc</span><span class="o">.</span><span class="n">html</span>                      <span class="mi">100</span><span class="o">%</span>  <span class="mi">251</span><span class="n">KB</span> <span class="mf">592.2</span><span class="n">KB</span><span class="o">/</span><span class="n">s</span>   <span class="mi">00</span><span class="p">:</span><span class="mi">00</span>
</pre></div>
</div>
<p>Now we can go to our new directory and open the HTML files.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd ~/Desktop/fastqc_html/
$ open *.html
</pre></div>
</div>
<p>Your computer will open each of the HTML files in your default web
browser. Depending on your settings, this might be as six separate
tabs in a single window or six separate browser windows.</p>
<blockquote>
<div><p><strong>Exercise</strong></p>
<p>Discuss your results with a neighbor. Which sample(s) looks the best
in terms of per base sequence quality? Which sample(s) look the
worst?</p>
<blockquote>
<div><em>Solution</em>
All of the reads contain usable data, but the quality decreases toward
the end of the reads.</div></blockquote>
</div></blockquote>
</div>
<div class="section" id="decoding-the-other-fastqc-outputs">
<h2>15.5. Decoding the other FastQC outputs<a class="headerlink" href="#decoding-the-other-fastqc-outputs" title="Permalink to this headline">¶</a></h2>
<p>We’ve now looked at quite a few “Per base sequence quality” FastQC graphs, but there are nine other graphs that we haven’t talked about! Below we have provided a brief overview of interpretations for each of these plots. It’s important to keep in mind</p>
<ul class="simple">
<li><strong>Per tile sequence quality</strong>: the machines that perform sequencing are divided into tiles. This plot displays patterns in base quality along these tiles. Consistently low scores are often found around the edges, but hot spots can also occur in the middle if an air bubble was introduced at some point during the run.</li>
<li><strong>Per sequence quality scores</strong>: a density plot of quality for all reads at all positions. This plot shows what quality scores are most common.</li>
<li><strong>Per base sequence content</strong>: plots the proportion of each base position over all of the reads. Typically, we expect to see each base roughly 25% of the time at each position, but this often fails at the beginning or end of the read due to quality or adapter content.</li>
<li><strong>Per sequence GC content</strong>: a density plot of average GC content in each of the reads.</li>
<li><strong>Per base N content</strong>: the percent of times that ‘N’ occurs at a position in all reads. If there is an increase at a particular position, this might indicate that something went wrong during sequencing.</li>
<li><strong>Sequence Length Distribution</strong>: the distribution of sequence lengths of all reads in the file. If the data is raw, there is often on sharp peak, however if the reads have been trimmed, there may be a distribution of shorter lengths.</li>
<li><strong>Sequence Duplication Levels</strong>: A distribution of duplicated sequences. In sequencing, we expect most reads to only occur once. If some sequences are occurring more than once, it might indicate enrichment bias (e.g. from PCR). If the samples are high coverage (or RNA-seq or amplicon), this might not be true.</li>
<li><strong>Overrepresented sequences</strong>: A list of sequences that occur more frequently than would be expected by chance.</li>
<li><strong>Adapter Content</strong>: a graph indicating where adapater sequences occur in the reads.</li>
</ul>
</div>
<div class="section" id="working-with-the-fastqc-text-output">
<h2>15.6. Working with the FastQC text output<a class="headerlink" href="#working-with-the-fastqc-text-output" title="Permalink to this headline">¶</a></h2>
<p>Now that we’ve looked at our HTML reports to get a feel for the data,
let’s look more closely at the other output files. Go back to the tab
in your terminal program that is connected to your AWS instance
(the tab lab will start with <code class="docutils literal notranslate"><span class="pre">dcuser&#64;ip</span></code>) and make sure you’re in
our results subdirectory.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd ~/dc_workshop/results/fastqc_untrimmed_reads/
$ ls
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">SRR2584863_1_fastqc</span><span class="o">.</span><span class="n">html</span>  <span class="n">SRR2584866_1_fastqc</span><span class="o">.</span><span class="n">html</span>  <span class="n">SRR2589044_1_fastqc</span><span class="o">.</span><span class="n">html</span>
<span class="n">SRR2584863_1_fastqc</span><span class="o">.</span><span class="n">zip</span>   <span class="n">SRR2584866_1_fastqc</span><span class="o">.</span><span class="n">zip</span>   <span class="n">SRR2589044_1_fastqc</span><span class="o">.</span><span class="n">zip</span>
<span class="n">SRR2584863_2_fastqc</span><span class="o">.</span><span class="n">html</span>  <span class="n">SRR2584866_2_fastqc</span><span class="o">.</span><span class="n">html</span>  <span class="n">SRR2589044_2_fastqc</span><span class="o">.</span><span class="n">html</span>
<span class="n">SRR2584863_2_fastqc</span><span class="o">.</span><span class="n">zip</span>   <span class="n">SRR2584866_2_fastqc</span><span class="o">.</span><span class="n">zip</span>   <span class="n">SRR2589044_2_fastqc</span><span class="o">.</span><span class="n">zip</span>
</pre></div>
</div>
<p>Our <code class="docutils literal notranslate"><span class="pre">.zip</span></code> files are compressed files. They each contain multiple
different types of output files for a single input FASTQ file. To
view the contents of a <code class="docutils literal notranslate"><span class="pre">.zip</span></code> file, we can use the program <code class="docutils literal notranslate"><span class="pre">unzip</span></code>
to decompress these files. Let’s try doing them all at once using a
wildcard.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ unzip *.zip
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Archive</span><span class="p">:</span>  <span class="n">SRR2584863_1_fastqc</span><span class="o">.</span><span class="n">zip</span>
<span class="n">caution</span><span class="p">:</span> <span class="n">filename</span> <span class="ow">not</span> <span class="n">matched</span><span class="p">:</span>  <span class="n">SRR2584863_2_fastqc</span><span class="o">.</span><span class="n">zip</span>
<span class="n">caution</span><span class="p">:</span> <span class="n">filename</span> <span class="ow">not</span> <span class="n">matched</span><span class="p">:</span>  <span class="n">SRR2584866_1_fastqc</span><span class="o">.</span><span class="n">zip</span>
<span class="n">caution</span><span class="p">:</span> <span class="n">filename</span> <span class="ow">not</span> <span class="n">matched</span><span class="p">:</span>  <span class="n">SRR2584866_2_fastqc</span><span class="o">.</span><span class="n">zip</span>
<span class="n">caution</span><span class="p">:</span> <span class="n">filename</span> <span class="ow">not</span> <span class="n">matched</span><span class="p">:</span>  <span class="n">SRR2589044_1_fastqc</span><span class="o">.</span><span class="n">zip</span>
<span class="n">caution</span><span class="p">:</span> <span class="n">filename</span> <span class="ow">not</span> <span class="n">matched</span><span class="p">:</span>  <span class="n">SRR2589044_2_fastqc</span><span class="o">.</span><span class="n">zip</span>
</pre></div>
</div>
<p>This didn’t work. We unzipped the first file and then got a warning
message for each of the other <code class="docutils literal notranslate"><span class="pre">.zip</span></code> files. This is because <code class="docutils literal notranslate"><span class="pre">unzip</span></code>
expects to get only one zip file as input. We could go through and
unzip each file one at a time, but this is very time consuming and
error-prone. Someday you may have 500 files to unzip!</p>
<p>A more efficient way is to use a <code class="docutils literal notranslate"><span class="pre">for</span></code> loop like we learned in the Shell Genomics lesson to iterate through all of
our <code class="docutils literal notranslate"><span class="pre">.zip</span></code> files. Let’s see what that looks like and then we’ll
discuss what we’re doing with each line of our loop.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ for filename in *.zip
&gt; do
&gt; unzip $filename
&gt; done
</pre></div>
</div>
<p>In this example, the input is six filenames (one filename for each of our <code class="docutils literal notranslate"><span class="pre">.zip</span></code> files).
Each time the loop iterates, it will assign a file name to the variable <code class="docutils literal notranslate"><span class="pre">filename</span></code>
and run the <code class="docutils literal notranslate"><span class="pre">unzip</span></code> command.
The first time throug&gt; h the loop,
<code class="docutils literal notranslate"><span class="pre">$filename</span></code> is <code class="docutils literal notranslate"><span class="pre">SRR2584863_1_fastqc.zip</span></code>.
The interpreter runs the command <code class="docutils literal notranslate"><span class="pre">unzip</span></code> on <code class="docutils literal notranslate"><span class="pre">SRR2584863_1_fastqc.zip</span></code>.
For the second iteration, <code class="docutils literal notranslate"><span class="pre">$filename</span></code> becomes
<code class="docutils literal notranslate"><span class="pre">SRR2584863_2_fastqc.zip</span></code>. This time, the shell runs <code class="docutils literal notranslate"><span class="pre">unzip</span></code> on <code class="docutils literal notranslate"><span class="pre">SRR2584863_2_fastqc.zip</span></code>.
It then repeats this process for the four other <code class="docutils literal notranslate"><span class="pre">.zip</span></code> files in our directory.</p>
<p>When we run our <code class="docutils literal notranslate"><span class="pre">for</span></code> loop, you will see output that starts like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Archive</span><span class="p">:</span>  <span class="n">SRR2589044_2_fastqc</span><span class="o">.</span><span class="n">zip</span>
   <span class="n">creating</span><span class="p">:</span> <span class="n">SRR2589044_2_fastqc</span><span class="o">/</span>
   <span class="n">creating</span><span class="p">:</span> <span class="n">SRR2589044_2_fastqc</span><span class="o">/</span><span class="n">Icons</span><span class="o">/</span>
   <span class="n">creating</span><span class="p">:</span> <span class="n">SRR2589044_2_fastqc</span><span class="o">/</span><span class="n">Images</span><span class="o">/</span>
  <span class="n">inflating</span><span class="p">:</span> <span class="n">SRR2589044_2_fastqc</span><span class="o">/</span><span class="n">Icons</span><span class="o">/</span><span class="n">fastqc_icon</span><span class="o">.</span><span class="n">png</span>
  <span class="n">inflating</span><span class="p">:</span> <span class="n">SRR2589044_2_fastqc</span><span class="o">/</span><span class="n">Icons</span><span class="o">/</span><span class="n">warning</span><span class="o">.</span><span class="n">png</span>
  <span class="n">inflating</span><span class="p">:</span> <span class="n">SRR2589044_2_fastqc</span><span class="o">/</span><span class="n">Icons</span><span class="o">/</span><span class="n">error</span><span class="o">.</span><span class="n">png</span>
  <span class="n">inflating</span><span class="p">:</span> <span class="n">SRR2589044_2_fastqc</span><span class="o">/</span><span class="n">Icons</span><span class="o">/</span><span class="n">tick</span><span class="o">.</span><span class="n">png</span>
  <span class="n">inflating</span><span class="p">:</span> <span class="n">SRR2589044_2_fastqc</span><span class="o">/</span><span class="n">summary</span><span class="o">.</span><span class="n">txt</span>
  <span class="n">inflating</span><span class="p">:</span> <span class="n">SRR2589044_2_fastqc</span><span class="o">/</span><span class="n">Images</span><span class="o">/</span><span class="n">per_base_quality</span><span class="o">.</span><span class="n">png</span>
  <span class="n">inflating</span><span class="p">:</span> <span class="n">SRR2589044_2_fastqc</span><span class="o">/</span><span class="n">Images</span><span class="o">/</span><span class="n">per_tile_quality</span><span class="o">.</span><span class="n">png</span>
  <span class="n">inflating</span><span class="p">:</span> <span class="n">SRR2589044_2_fastqc</span><span class="o">/</span><span class="n">Images</span><span class="o">/</span><span class="n">per_sequence_quality</span><span class="o">.</span><span class="n">png</span>
  <span class="n">inflating</span><span class="p">:</span> <span class="n">SRR2589044_2_fastqc</span><span class="o">/</span><span class="n">Images</span><span class="o">/</span><span class="n">per_base_sequence_content</span><span class="o">.</span><span class="n">png</span>
  <span class="n">inflating</span><span class="p">:</span> <span class="n">SRR2589044_2_fastqc</span><span class="o">/</span><span class="n">Images</span><span class="o">/</span><span class="n">per_sequence_gc_content</span><span class="o">.</span><span class="n">png</span>
  <span class="n">inflating</span><span class="p">:</span> <span class="n">SRR2589044_2_fastqc</span><span class="o">/</span><span class="n">Images</span><span class="o">/</span><span class="n">per_base_n_content</span><span class="o">.</span><span class="n">png</span>
  <span class="n">inflating</span><span class="p">:</span> <span class="n">SRR2589044_2_fastqc</span><span class="o">/</span><span class="n">Images</span><span class="o">/</span><span class="n">sequence_length_distribution</span><span class="o">.</span><span class="n">png</span>
  <span class="n">inflating</span><span class="p">:</span> <span class="n">SRR2589044_2_fastqc</span><span class="o">/</span><span class="n">Images</span><span class="o">/</span><span class="n">duplication_levels</span><span class="o">.</span><span class="n">png</span>
  <span class="n">inflating</span><span class="p">:</span> <span class="n">SRR2589044_2_fastqc</span><span class="o">/</span><span class="n">Images</span><span class="o">/</span><span class="n">adapter_content</span><span class="o">.</span><span class="n">png</span>
  <span class="n">inflating</span><span class="p">:</span> <span class="n">SRR2589044_2_fastqc</span><span class="o">/</span><span class="n">fastqc_report</span><span class="o">.</span><span class="n">html</span>
  <span class="n">inflating</span><span class="p">:</span> <span class="n">SRR2589044_2_fastqc</span><span class="o">/</span><span class="n">fastqc_data</span><span class="o">.</span><span class="n">txt</span>
  <span class="n">inflating</span><span class="p">:</span> <span class="n">SRR2589044_2_fastqc</span><span class="o">/</span><span class="n">fastqc</span><span class="o">.</span><span class="n">fo</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">unzip</span></code> program is decompressing the <code class="docutils literal notranslate"><span class="pre">.zip</span></code> files and creating
a new directory (with subdirectories) for each of our samples, to
store all of the different output that is produced by FastQC. There
are a lot of files here. The one we’re going to focus on is the
<code class="docutils literal notranslate"><span class="pre">summary.txt</span></code> file.</p>
<p>If you list the files in our directory now you will see:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">SRR2584863_1_fastqc</span>       <span class="n">SRR2584866_1_fastqc</span>       <span class="n">SRR2589044_1_fastqc</span>
<span class="n">SRR2584863_1_fastqc</span><span class="o">.</span><span class="n">html</span>  <span class="n">SRR2584866_1_fastqc</span><span class="o">.</span><span class="n">html</span>  <span class="n">SRR2589044_1_fastqc</span><span class="o">.</span><span class="n">html</span>
<span class="n">SRR2584863_1_fastqc</span><span class="o">.</span><span class="n">zip</span>   <span class="n">SRR2584866_1_fastqc</span><span class="o">.</span><span class="n">zip</span>   <span class="n">SRR2589044_1_fastqc</span><span class="o">.</span><span class="n">zip</span>
<span class="n">SRR2584863_2_fastqc</span>       <span class="n">SRR2584866_2_fastqc</span>       <span class="n">SRR2589044_2_fastqc</span>
<span class="n">SRR2584863_2_fastqc</span><span class="o">.</span><span class="n">html</span>  <span class="n">SRR2584866_2_fastqc</span><span class="o">.</span><span class="n">html</span>  <span class="n">SRR2589044_2_fastqc</span><span class="o">.</span><span class="n">html</span>
<span class="n">SRR2584863_2_fastqc</span><span class="o">.</span><span class="n">zip</span>   <span class="n">SRR2584866_2_fastqc</span><span class="o">.</span><span class="n">zip</span>   <span class="n">SRR2589044_2_fastqc</span><span class="o">.</span><span class="n">zip</span>
</pre></div>
</div>
<p>{:. output}</p>
<p>The <code class="docutils literal notranslate"><span class="pre">.html</span></code> files and the uncompressed <code class="docutils literal notranslate"><span class="pre">.zip</span></code> files are still present,
but now we also have a new directory for each of our samples. We can
see for sure that it’s a directory if we use the <code class="docutils literal notranslate"><span class="pre">-F</span></code> flag for <code class="docutils literal notranslate"><span class="pre">ls</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ls -F
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">SRR2584863_1_fastqc</span><span class="o">/</span>      <span class="n">SRR2584866_1_fastqc</span><span class="o">/</span>      <span class="n">SRR2589044_1_fastqc</span><span class="o">/</span>
<span class="n">SRR2584863_1_fastqc</span><span class="o">.</span><span class="n">html</span>  <span class="n">SRR2584866_1_fastqc</span><span class="o">.</span><span class="n">html</span>  <span class="n">SRR2589044_1_fastqc</span><span class="o">.</span><span class="n">html</span>
<span class="n">SRR2584863_1_fastqc</span><span class="o">.</span><span class="n">zip</span>   <span class="n">SRR2584866_1_fastqc</span><span class="o">.</span><span class="n">zip</span>   <span class="n">SRR2589044_1_fastqc</span><span class="o">.</span><span class="n">zip</span>
<span class="n">SRR2584863_2_fastqc</span><span class="o">/</span>      <span class="n">SRR2584866_2_fastqc</span><span class="o">/</span>      <span class="n">SRR2589044_2_fastqc</span><span class="o">/</span>
<span class="n">SRR2584863_2_fastqc</span><span class="o">.</span><span class="n">html</span>  <span class="n">SRR2584866_2_fastqc</span><span class="o">.</span><span class="n">html</span>  <span class="n">SRR2589044_2_fastqc</span><span class="o">.</span><span class="n">html</span>
<span class="n">SRR2584863_2_fastqc</span><span class="o">.</span><span class="n">zip</span>   <span class="n">SRR2584866_2_fastqc</span><span class="o">.</span><span class="n">zip</span>   <span class="n">SRR2589044_2_fastqc</span><span class="o">.</span><span class="n">zip</span>
</pre></div>
</div>
<p>Let’s see what files are present within one of these output directories.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ls -F SRR2584863_1_fastqc/
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">fastqc_data</span><span class="o">.</span><span class="n">txt</span>  <span class="n">fastqc</span><span class="o">.</span><span class="n">fo</span>  <span class="n">fastqc_report</span><span class="o">.</span><span class="n">html</span>	<span class="n">Icons</span><span class="o">/</span>	<span class="n">Images</span><span class="o">/</span>  <span class="n">summary</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>Use <code class="docutils literal notranslate"><span class="pre">less</span></code> to preview the <code class="docutils literal notranslate"><span class="pre">summary.txt</span></code> file for this sample.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ less SRR2584863_1_fastqc/summary.txt
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">PASS</span>    <span class="n">Basic</span> <span class="n">Statistics</span>        <span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">PASS</span>    <span class="n">Per</span> <span class="n">base</span> <span class="n">sequence</span> <span class="n">quality</span>       <span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">PASS</span>    <span class="n">Per</span> <span class="n">tile</span> <span class="n">sequence</span> <span class="n">quality</span>       <span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">PASS</span>    <span class="n">Per</span> <span class="n">sequence</span> <span class="n">quality</span> <span class="n">scores</span>     <span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">WARN</span>    <span class="n">Per</span> <span class="n">base</span> <span class="n">sequence</span> <span class="n">content</span>       <span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">WARN</span>    <span class="n">Per</span> <span class="n">sequence</span> <span class="n">GC</span> <span class="n">content</span> <span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">PASS</span>    <span class="n">Per</span> <span class="n">base</span> <span class="n">N</span> <span class="n">content</span>      <span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">PASS</span>    <span class="n">Sequence</span> <span class="n">Length</span> <span class="n">Distribution</span>    <span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">PASS</span>    <span class="n">Sequence</span> <span class="n">Duplication</span> <span class="n">Levels</span>     <span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">PASS</span>    <span class="n">Overrepresented</span> <span class="n">sequences</span>       <span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">WARN</span>    <span class="n">Adapter</span> <span class="n">Content</span> <span class="n">SRR2584863_1</span><span class="o">.</span><span class="n">fastq</span>
</pre></div>
</div>
<p>The summary file gives us a list of tests that FastQC ran, and tells
us whether this sample passed, failed, or is borderline (<code class="docutils literal notranslate"><span class="pre">WARN</span></code>). Remember to quit from <code class="docutils literal notranslate"><span class="pre">less</span></code> you enter <code class="docutils literal notranslate"><span class="pre">q</span></code>.</p>
</div>
<div class="section" id="documenting-our-work">
<h2>15.7. Documenting Our Work<a class="headerlink" href="#documenting-our-work" title="Permalink to this headline">¶</a></h2>
<p>We can make a record of the results we obtained for all our samples
by concatenating all of our <code class="docutils literal notranslate"><span class="pre">summary.txt</span></code> files into a single file
using the <code class="docutils literal notranslate"><span class="pre">cat</span></code> command. We’ll call this <code class="docutils literal notranslate"><span class="pre">full_report.txt</span></code> and move
it to <code class="docutils literal notranslate"><span class="pre">~/dc_workshop/docs</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cat */summary.txt &gt; ~/dc_workshop/docs/fastqc_summaries.txt
</pre></div>
</div>
<blockquote>
<div><p><strong>Exercise</strong></p>
<p>Which samples failed at least one of FastQC’s quality tests? What
test(s) did those samples fail?</p>
<blockquote>
<div><p><em>Solution</em></p>
<p>We can get the list of all failed tests using <code class="docutils literal notranslate"><span class="pre">grep</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd ~/dc_workshop/docs
$ grep FAIL fastqc_summaries.txt
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">FAIL</span>    <span class="n">Per</span> <span class="n">base</span> <span class="n">sequence</span> <span class="n">quality</span>       <span class="n">SRR2584863_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">FAIL</span>    <span class="n">Per</span> <span class="n">tile</span> <span class="n">sequence</span> <span class="n">quality</span>       <span class="n">SRR2584863_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">FAIL</span>    <span class="n">Per</span> <span class="n">base</span> <span class="n">sequence</span> <span class="n">content</span>       <span class="n">SRR2584863_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">FAIL</span>    <span class="n">Per</span> <span class="n">base</span> <span class="n">sequence</span> <span class="n">quality</span>       <span class="n">SRR2584866_1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">FAIL</span>    <span class="n">Per</span> <span class="n">base</span> <span class="n">sequence</span> <span class="n">content</span>       <span class="n">SRR2584866_1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">FAIL</span>    <span class="n">Adapter</span> <span class="n">Content</span> <span class="n">SRR2584866_1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">FAIL</span>    <span class="n">Adapter</span> <span class="n">Content</span> <span class="n">SRR2584866_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">FAIL</span>    <span class="n">Adapter</span> <span class="n">Content</span> <span class="n">SRR2589044_1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">FAIL</span>    <span class="n">Per</span> <span class="n">base</span> <span class="n">sequence</span> <span class="n">quality</span>       <span class="n">SRR2589044_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">FAIL</span>    <span class="n">Per</span> <span class="n">tile</span> <span class="n">sequence</span> <span class="n">quality</span>       <span class="n">SRR2589044_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">FAIL</span>    <span class="n">Per</span> <span class="n">base</span> <span class="n">sequence</span> <span class="n">content</span>       <span class="n">SRR2589044_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">FAIL</span>    <span class="n">Adapter</span> <span class="n">Content</span> <span class="n">SRR2589044_2</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</div></blockquote>
</div></blockquote>
</div>
</div>
<div class="section" id="other-notes-optional">
<h1>16. Other notes  – Optional<a class="headerlink" href="#other-notes-optional" title="Permalink to this headline">¶</a></h1>
<blockquote>
<div><p><strong>Quality Encodings Vary</strong></p>
<p>Although we’ve used a particular quality encoding system to demonstrate interpretation of
read quality, different sequencing machines use different encoding systems. This means that,
depending on which sequencer you use to generate your data, a <code class="docutils literal notranslate"><span class="pre">#</span></code> may not be an indicator of
a poor quality base call.</p>
<p>This mainly relates to older Solexa/Illumina data,
but it’s essential that you know which sequencing platform was
used to generate your data, so that you can tell your quality control program which encoding
to use. If you choose the wrong encoding, you run the risk of throwing away good reads or
(even worse) not throwing away bad reads!</p>
</div></blockquote>
<blockquote>
<div><p><strong>Same Symbols, Different Meanings</strong></p>
<p>Here we see <code class="docutils literal notranslate"><span class="pre">&gt;</span></code> being used a shell prompt, whereas <code class="docutils literal notranslate"><span class="pre">&gt;</span></code> is also
used to redirect output.
Similarly, <code class="docutils literal notranslate"><span class="pre">$</span></code> is used as a shell prompt, but, as we saw earlier,
it is also used to ask the shell to get the value of a variable.</p>
<p>If the <em>shell</em> prints <code class="docutils literal notranslate"><span class="pre">&gt;</span></code> or <code class="docutils literal notranslate"><span class="pre">$</span></code> then it expects you to type something,
and the symbol is a prompt.</p>
<p>If <em>you</em> type <code class="docutils literal notranslate"><span class="pre">&gt;</span></code> or <code class="docutils literal notranslate"><span class="pre">$</span></code> yourself, it is an instruction from you that
the shell to redirect output or get the value of a variable.</p>
</div></blockquote>
</div>


          </div>
            
  <div class="footer-relations">
    
      <div class="pull-left">
        <a class="btn btn-default" href="bioconda_lesson.html" title="previous chapter (use the left arrow)">11. Working with Bioconda.</a>
      </div>
    
      <div class="pull-right">
        <a class="btn btn-default" href="trimming.html" title="next chapter (use the right arrow)">17. Trimming and Filtering</a>
      </div>
    </div>
    <div class="clearer"></div>
  
        </div>
        <div class="clearfix"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="trimming.html" title="17. Trimming and Filtering"
             >next</a> |</li>
        <li class="right" >
          <a href="bioconda_lesson.html" title="11. Working with Bioconda."
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="toc.html">2019-04-11-gwu 0.0.1 documentation</a> &#187;</li> 
      </ul>
    </div>
<script type="text/javascript">
  $("#mobile-toggle a").click(function () {
    $("#left-column").toggle();
  });
</script>
<script type="text/javascript" src="_static/js/bootstrap.js"></script>
  <div class="footer">
    &copy; Copyright 2019, Sateesh Peri. Created using <a href="http://sphinx.pocoo.org/">Sphinx</a>.
  </div>
  </body>
</html>